# **Business Performance & Financial Planning Project**

## Project Background

### Business Context

**Texas Rex Inc.** is a manufacturing company specializing in the production of high-quality logoed T-shirts.

This project was inspired by the **Master Budget** case study presented in *Managerial Accounting: The Cornerstone of Business Decision Making (7th Edition)* by **Mowen, Hansen, and Heitger**. The original budgeting framework has been redesigned and extended into an automated financial planning and business performance analysis solution using **Python, Excel VBA, and Power BI**.

To support sustainable growth and improve operational decision-making, management plans to implement an integrated budgeting and business performance reporting process for the upcoming fiscal year.

Historically, the company's budgeting activities were prepared manually using spreadsheets, making it difficult to:

- Monitor business performance consistently.
- Evaluate operational efficiency.
- Identify cost drivers.
- Perform timely financial analysis.
- Support management decision-making.

As the business continues to expand, management requires a more structured, automated, and data-driven financial planning process.

---

## Business Problem

Management requires an integrated budgeting and reporting solution capable of:

- Forecasting future business performance.
- Supporting annual budgeting and financial planning.
- Monitoring operational and financial KPIs.
- Evaluating cost efficiency and profitability.
- Improving reporting accuracy and consistency.
- Supporting strategic management decision-making.

In addition, management aims to automate the budgeting workflow to minimize manual calculations, reduce reporting errors, and improve overall reporting efficiency.

---

## Project Objective

The objective of this project is to develop an end-to-end **Master Budget Automation & Financial Performance Analysis** solution using Python.

The solution automates the preparation of annual operating budgets, projected financial statements, financial performance analysis, and executive reporting within a single integrated workflow.

---

## Business Scope

The project covers the complete Master Budget process, including:

- Sales Budget
- Production Budget
- Direct Materials Purchases Budget
- Direct Labor Budget
- Manufacturing Overhead Budget
- Ending Finished Goods Inventory Budget
- Cost of Goods Sold Budget
- Selling & Administrative Expense Budget
- Cash Collections Budget
- Cash Disbursements Budget
- Cash Budget
- Budgeted Income Statement
- Budgeted Statement of Financial Position (SOFP)
- Financial Ratio Analysis
- Vertical & Horizontal Analysis
- Altman Z-Score Analysis
- Executive Dashboard

---

## Business Value

The proposed solution enables management to:

- Improve budgeting accuracy and consistency across business functions.
- Support annual budgeting and financial forecasting.
- Monitor financial and operational performance through automated reporting.
- Analyze cost structures, profitability, liquidity, solvency, and cash flow.
- Evaluate overall financial health using key financial indicators.
- Support strategic planning through data-driven business insights.
- Reduce manual spreadsheet processing by automating budgeting and reporting workflows.

---

## Executive Dashboard

To support executive reporting and business performance monitoring, the final financial outputs are transformed into an interactive **Power BI Executive Dashboard**.

The dashboard provides management with a consolidated view of projected financial performance through:

- Executive KPI Scorecards
- Budgeted Income Statement Summary
- Budgeted Statement of Financial Position Summary
- Cash Budget Overview
- Financial Ratio Analysis
- Altman Z-Score
- Quarterly Performance Trends
- Revenue & Cost Analysis
- Profitability Analysis

The dashboard enables management to monitor business performance, identify financial trends, evaluate key performance indicators, and support strategic decision-making through interactive visual analytics.

In [ ]:
import pandas as pd
filename = 'Budgeting Case_portfolio.xlsx'
Budgeting = pd.read_excel(filename, sheet_name ='Data')
Budgeting


,Account,1,2,3,4
0,Sales Unit,1000.00,1200.00,1500.00,2000.00
1,Sell Prices,10.00,10.00,10.00,10.00
2,Beginning Inventory,180.00,240.00,300.00,400.00
3,Ending Inventory,240.00,300.00,400.00,200.00
4,Beginning Inventory Direct Material A,58.00,126.00,160.00,180.00
5,Ending Inventory Direct Material A,126.00,160.00,180.00,106.00
6,Beginning Inventory Direct Material B,390.00,630.00,800.00,900.00
7,Ending Inventory Direct Material B,630.00,800.00,900.00,530.00
8,Price Direct material A Raw Material,1.00,1.00,1.00,1.00
9,Price Direct Material B Raw Material,5.00,5.00,5.00,5.00


# **Sales Bugdeting**


### Business Scenario

**Texas Rex Apparel** is preparing its **Annual Master Budget** for the upcoming fiscal year. As the first step in the budgeting process, management must develop a **Sales Budget**, which serves as the foundation for all subsequent operating and financial budgets.

Based on market research, historical sales performance, and management's sales forecast, the company expects the following sales volume for the upcoming year:

| Quarter | Budgeted Unit Sales |
|:--------|--------------------:|
| Q1 | 1,000 |
| Q2 | 1,200 |
| Q3 | 1,500 |
| Q4 | 2,000 |

The company sells each customized T-shirt at a **selling price of $10 per unit**.

The Sales Budget will be used to estimate quarterly sales revenue and provide the basis for preparing the Production Budget, Cash Budget, and Budgeted Financial Statements.

---

## Required

Prepare the **Sales Budget** for each quarter and for the fiscal year.

In [ ]:

#Sales Budgeting
#Source Data
Sales_Unit = Budgeting.loc[Budgeting['Account'] == 'Sales Unit'].iloc[:,1:]
Price_Sell = Budgeting.loc[Budgeting['Account'] == 'Sell Prices'].iloc[:,1:]
#Calculation
Budgeted_Sales = Price_Sell.values * Sales_Unit.values
#Data Frame
Quarter = Budgeting.columns [1:]
df_Sales_Unit = pd.DataFrame(Sales_Unit.values, columns = Quarter, index = ['Unit'])
df_Price_Sell = pd.DataFrame(Price_Sell.values, columns = Quarter, index = ['Price Sell'])
df_Budgeted_Sales = pd.DataFrame(Budgeted_Sales, columns = Quarter, index = ['Budgeted Sales'])

#Table
df_Sales_Budget = pd.concat([df_Sales_Unit, df_Price_Sell, df_Budgeted_Sales])

#New Column year
df_Sales_Budget['year'] = df_Sales_Budget.apply(lambda row : row.mean() if row.name == 'Price Sell' else row.sum(), axis = 1)

#Title
styled_df = ( df_Sales_Budget.style
             .set_caption("Sales Budget")
            )
styled_df


,1,2,3,4,year
Unit,1000.000000,1200.000000,1500.000000,2000.000000,5700.000000
Price Sell,10.000000,10.000000,10.000000,10.000000,10.000000
Budgeted Sales,10000.000000,12000.000000,15000.000000,20000.000000,57000.000000


# **Production Budget**

### Business Scenario

Following the completion of the **Sales Budget**, **Texas Rex Apparel** must determine the number of T-shirts to produce during each quarter of the upcoming fiscal year.

To ensure uninterrupted production while minimizing inventory holding costs, management has established an inventory policy requiring the company to maintain an ending inventory equal to **20% of the following quarter's expected unit sales**.

At the beginning of the budget year, the company has **180 finished T-shirts** in inventory. In addition, management forecasts that **first-quarter sales for the following fiscal year will be 1,000 units**, which will be used to determine the desired ending inventory for the fourth quarter.

The Production Budget will estimate the number of units that must be manufactured each quarter to satisfy customer demand while maintaining the company's inventory policy.

---

## Required

Using the information provided in the **Sales Budget**:

1. Calculate the **desired ending finished goods inventory** for each quarter and for the fiscal year.

2. Prepare the **Production Budget** for each quarter and for the fiscal year.

In [ ]:

#Production Budget

#Source
Beginning_Inventory = Budgeting.loc[Budgeting['Account'] == 'Beginning Inventory'].iloc[:,1:]
Ending_Inventory = Budgeting.loc[Budgeting['Account'] == 'Ending Inventory'].iloc[:,1:]

#Calculation

Total_Needs = Sales_Unit.values + Ending_Inventory.values
Unit_to_be_Produced = Total_Needs - Beginning_Inventory.values

#DataFrame

df_Beginning_Inventory = pd.DataFrame(Beginning_Inventory.values, columns = Quarter, index = ['Beginning Inventory'])
df_Ending_Inventory = pd.DataFrame(Ending_Inventory.values, columns = Quarter, index = ['Ending Inventory'])
df_Total_Needs = pd.DataFrame(Total_Needs, columns = Quarter, index = [' Total Needs'] )
df_Unit_to_be_Produced = pd.DataFrame(Unit_to_be_Produced, columns = Quarter, index = ['Unit to be Produced'])

#Table
df_Production_Budget = pd.concat([df_Sales_Unit,df_Beginning_Inventory,df_Total_Needs,df_Ending_Inventory,df_Unit_to_be_Produced])

#Year Column
df_Production_Budget['year'] = df_Sales_Budget.apply(lambda row : row.mean() if row.name == 'Price Sell' else row.sum(), axis = 1)

#Title
styled_df = ( df_Sales_Budget.style
             .set_caption("Production Budget")
            )
styled_df


,1,2,3,4,year
Unit,1000.000000,1200.000000,1500.000000,2000.000000,5700.000000
Price Sell,10.000000,10.000000,10.000000,10.000000,10.000000
Budgeted Sales,10000.000000,12000.000000,15000.000000,20000.000000,57000.000000


# **Direct Material**

### Business Scenario

Following the completion of the **Production Budget**, **Texas Rex Apparel** must estimate the quantity and cost of direct materials required to support its planned production for the upcoming fiscal year.

Each customized T-shirt requires the following direct materials:

- **1 plain T-shirt**, costing **$3.00 per unit**
- **5 ounces of ink**, costing **$0.20 per ounce**

To ensure uninterrupted production while controlling inventory carrying costs, management has adopted an inventory policy requiring ending raw materials inventory to equal **10% of the following quarter's production requirements**.

At the beginning of the budget year (**January 1**), the company has the following raw materials on hand:

| Direct Material | Beginning Inventory |
|:----------------|--------------------:|
| Plain T-shirts | 58 units |
| Ink | 390 ounces |

To support production planning beyond the current fiscal year, management also establishes the following desired ending inventory at **December 31**:

| Direct Material | Desired Ending Inventory |
|:----------------|-------------------------:|
| Plain T-shirts | 106 units |
| Ink | 530 ounces |

The Direct Materials Purchases Budget will determine the quantity of materials that must be purchased during each quarter and estimate the total purchasing cost required to meet production needs.

---

## Required

Using the **Production Budget** and the information provided above:

1. Calculate the desired ending inventory of **plain T-shirts** and **ink** for each quarter.

2. Prepare the **Direct Materials Purchases Budget** for:
   - Plain T-shirts
   - Ink

3. Determine the total quantity and cost of direct materials to be purchased for each quarter and for the fiscal year.

# **Direct Material A (T-Shirt)**

In [ ]:

#Direct Material
#Source Data
Beginning_Inventory_DM_A = Budgeting.loc[Budgeting['Account'] == 'Beginning Inventory Direct Material A'].iloc[:,1:]
Ending_Inventory_DM_A = Budgeting.loc[Budgeting['Account'] == 'Ending Inventory Direct Material A'].iloc[:,1:]
Beginning_Inventory_DM_B = Budgeting.loc[Budgeting['Account'] == 'Beginning Inventory Direct Material B'].iloc[:,1:]
Ending_Inventory_DM_B = Budgeting.loc[Budgeting['Account'] == 'Ending Inventory Direct Material B'].iloc[:,1:]
Price_DM_A_Raw = Budgeting.loc[Budgeting['Account'] == 'Price Direct material A Raw Material'].iloc[:,1:]
Price_DM_A_FG = Budgeting.loc[Budgeting['Account'] == 'Direct material A cost (finish Good)'].iloc[:,1:]
Price_DM_B_Raw = Budgeting.loc[Budgeting['Account'] == 'Price  Direct Material B Raw Material'].iloc[:,1:]
Price_DM_B_FG = Budgeting.loc[Budgeting['Account'] == 'Direct material B cost (finish Good)'].iloc[:,1:]

#Direct Material A

#Calculation
Production_Need_DM_A = Unit_to_be_Produced * Price_DM_A_Raw.values
Total_Needs_DM_A = Production_Need_DM_A + Ending_Inventory_DM_A.values
DM_A_Purchase_Estimated = Total_Needs_DM_A - Beginning_Inventory_DM_A.values
Direct_Material_A = DM_A_Purchase_Estimated * Price_DM_A_FG.values

#DataFrame
df_Price_DM_A_Raw = pd.DataFrame(Price_DM_A_Raw.values, columns = Quarter, index = ['Price DM A Raw'])
df_Production_Need_DM_A = pd.DataFrame(Production_Need_DM_A, columns = Quarter, index = ['Production Need DM A'])
df_Ending_Inventory_DM_A = pd.DataFrame(Ending_Inventory_DM_A.values, columns = Quarter, index = ['Ending Inventory DM A'])
df_Total_Needs_DM_A = pd.DataFrame(Total_Needs_DM_A, columns = Quarter, index = ['Total Needs DM A'])
df_Beginning_Inventory_DM_A = pd.DataFrame(Beginning_Inventory_DM_A.values, columns = Quarter, index = ['Beginning Inventory DM A'])
df_Direct_Material_A = pd.DataFrame(Direct_Material_A, columns = Quarter, index = ['Direct Material A'])
df_DM_A_Purchase_Estimated = pd.DataFrame(DM_A_Purchase_Estimated, columns = Quarter, index = ['DM A Purchase Estimated'])
df_Price_DM_A_FG = pd.DataFrame(Price_DM_A_FG.values, columns = Quarter, index = ['Price DM A FG'])

#Table Direct Material
df_Direct_Material_A_Budget = pd.concat([df_Unit_to_be_Produced,df_Price_DM_A_Raw,df_Production_Need_DM_A,df_Ending_Inventory_DM_A, df_Total_Needs_DM_A, df_Beginning_Inventory_DM_A, df_DM_A_Purchase_Estimated,df_Price_DM_A_FG,df_Direct_Material_A])
#Year Columns

#Year Column
df_Direct_Material_A_Budget['year'] = df_Direct_Material_A_Budget.apply(lambda row : row.mean() if row.name == ['Price DM A Raw','Price DM A FG'] else row.sum(), axis = 1)

#Title
styled_df = ( df_Direct_Material_A_Budget.style
             .set_caption("Direct Material A Budget")
   	     )
styled_df


,1,2,3,4,year
Unit to be Produced,1060.000000,1260.000000,1600.000000,1800.000000,5720.000000
Price DM A Raw,1.000000,1.000000,1.000000,1.000000,4.000000
Production Need DM A,1060.000000,1260.000000,1600.000000,1800.000000,5720.000000
Ending Inventory DM A,126.000000,160.000000,180.000000,106.000000,572.000000
Total Needs DM A,1186.000000,1420.000000,1780.000000,1906.000000,6292.000000
Beginning Inventory DM A,58.000000,126.000000,160.000000,180.000000,524.000000
DM A Purchase Estimated,1128.000000,1294.000000,1620.000000,1726.000000,5768.000000
Price DM A FG,3.000000,3.000000,3.000000,3.000000,12.000000
Direct Material A,3384.000000,3882.000000,4860.000000,5178.000000,17304.000000


# **Direct Material B (Ink)**

In [ ]:
# --- DIRECT MATERIAL B ---

# Calculation
Production_Need_DM_B = Unit_to_be_Produced * Price_DM_B_Raw.values
Total_Needs_DM_B = Production_Need_DM_B + Ending_Inventory_DM_B.values
DM_B_Purchase_Estimated = Total_Needs_DM_B - Beginning_Inventory_DM_B.values
Direct_Material_B = DM_B_Purchase_Estimated * Price_DM_B_FG.values

# DataFrame
df_Price_DM_B_Raw = pd.DataFrame(Price_DM_B_Raw.values, columns=Quarter, index=['Price DM B Raw'])
df_Production_Need_DM_B = pd.DataFrame(Production_Need_DM_B, columns=Quarter, index=['Production Need DM B'])
df_Ending_Inventory_DM_B = pd.DataFrame(Ending_Inventory_DM_B.values, columns=Quarter, index=['Ending Inventory DM B'])
df_Total_Needs_DM_B = pd.DataFrame(Total_Needs_DM_B, columns=Quarter, index=['Total Needs DM B'])
df_Beginning_Inventory_DM_B = pd.DataFrame(Beginning_Inventory_DM_B.values, columns=Quarter, index=['Beginning Inventory DM B'])
df_Direct_Material_B = pd.DataFrame(Direct_Material_B, columns=Quarter, index=['Direct Material B'])
df_DM_B_Purchase_Estimated = pd.DataFrame(DM_B_Purchase_Estimated, columns=Quarter, index=['DM B Purchase Estimated'])
df_Price_DM_B_FG = pd.DataFrame(Price_DM_B_FG.values, columns=Quarter, index=['Price DM B FG'])

# Table Direct Material B
df_Direct_Material_B_Budget = pd.concat([
    df_Unit_to_be_Produced,
    df_Price_DM_B_Raw,
    df_Production_Need_DM_B,
    df_Ending_Inventory_DM_B,
    df_Total_Needs_DM_B,
    df_Beginning_Inventory_DM_B,
    df_DM_B_Purchase_Estimated,
    df_Price_DM_B_FG,
    df_Direct_Material_B
])

# Year Columns
df_Direct_Material_B_Budget['year'] = df_Direct_Material_B_Budget.apply(
    lambda row: row.mean() if row.name in ['Price DM B Raw', 'Price DM B FG'] else row.sum(),
    axis=1
)

# Title
styled_df_B = (
    df_Direct_Material_B_Budget.style
    .set_caption("Direct Material B Budget")
)

styled_df_B


,1,2,3,4,year
Unit to be Produced,1060.000000,1260.000000,1600.000000,1800.000000,5720.000000
Price DM B Raw,5.000000,5.000000,5.000000,5.000000,5.000000
Production Need DM B,5300.000000,6300.000000,8000.000000,9000.000000,28600.000000
Ending Inventory DM B,630.000000,800.000000,900.000000,530.000000,2860.000000
Total Needs DM B,5930.000000,7100.000000,8900.000000,9530.000000,31460.000000
Beginning Inventory DM B,390.000000,630.000000,800.000000,900.000000,2720.000000
DM B Purchase Estimated,5540.000000,6470.000000,8100.000000,8630.000000,28740.000000
Price DM B FG,0.200000,0.200000,0.200000,0.200000,0.200000
Direct Material B,1108.000000,1294.000000,1620.000000,1726.000000,5748.000000


# **Total Direct Material**

In [ ]:
#Total Direct Material

#Calculation
Total_Direct_Material = Direct_Material_A + Direct_Material_B
#DataFrame
df_Total_Direct_Material = pd.DataFrame(Total_Direct_Material, columns = Quarter, index = ['Total Direct Material'])

#year Columns
df_Total_Direct_Material['year'] = df_Total_Direct_Material.apply(lambda row : row.sum(), axis = 1)


#Title
styled_df = ( df_Total_Direct_Material.style
             .set_caption("Total Direct Material Budget")
            )
styled_df


,1,2,3,4,year
Total Direct Material,4492.000000,5176.000000,6480.000000,6904.000000,23052.000000


# **Direct Labor**

### Business Scenario

Following the completion of the **Production Budget**, **Texas Rex Apparel** must estimate the direct labor resources required to meet its planned production schedule for the upcoming fiscal year.

Based on the Production Budget, the expected production volume is as follows:

| Quarter | Budgeted Production |
|:--------|--------------------:|
| Q1 | 1,060 units |
| Q2 | 1,260 units |
| Q3 | 1,600 units |
| Q4 | 1,800 units |

The production process requires **0.12 direct labor hours** to manufacture one customized T-shirt. The company pays an average wage of **$10.00 per direct labor hour** for all production employees.

The Direct Labor Budget will estimate the total labor hours required and the corresponding labor cost for each quarter. This budget also serves as an important input for preparing the **Manufacturing Overhead Budget** and forecasting total manufacturing costs.

---

## Required

Using the **Production Budget** and the labor cost assumptions provided above:

1. Calculate the total **direct labor hours required** for each quarter and for the fiscal year.

2. Prepare the **Direct Labor Budget**, including:
   - Total production units
   - Direct labor hours required
   - Direct labor cost per hour
   - Total direct labor cost for each quarter and for the fiscal year

In [ ]:
#Direct Labour
#Source
Direct_Labour_Hour = Budgeting.loc[Budgeting['Account'] == 'Direct labour Hour'].iloc[:,1:]
Wage_Labor_Hour = Budgeting.loc[Budgeting['Account'] == 'Wage labor hour'].iloc[:,1:]
#Calculation
Total_Hour_Needs = Unit_to_be_Produced * Direct_Labour_Hour.values
Total_Direct_Labor_cost = Total_Hour_Needs * Wage_Labor_Hour.values

#DataFrame
# DataFrame
df_Direct_Labour_Hour = pd.DataFrame(Direct_Labour_Hour.values, columns=Quarter, index=['Direct Labour Hour'])
df_Total_Hour_Needs = pd.DataFrame(Total_Hour_Needs, columns=Quarter, index=['Total Hour Needs'])
df_Wage_Labor_Hour = pd.DataFrame(Wage_Labor_Hour.values, columns=Quarter, index=['Wage Labor Hour'])
df_Total_Direct_Labor_Cost = pd.DataFrame(Total_Direct_Labor_cost, columns=Quarter, index=['Total Direct Labor Cost'])

# Table Direct Labour
df_Direct_Labor_Budget = pd.concat([
    df_Unit_to_be_Produced,
    df_Direct_Labour_Hour,
    df_Total_Hour_Needs,
    df_Wage_Labor_Hour,
    df_Total_Direct_Labor_Cost
])

# Year Columns
df_Direct_Labor_Budget['year'] = df_Direct_Labor_Budget.apply(
    lambda row: row.mean() if row.name in ['Direct Labour Hour', 'Wage Labor Hour'] else row.sum(),
    axis=1
)

# Title
styled_df_Labor = (
    df_Direct_Labor_Budget.style
    .set_caption("Direct Labour Budget")
)

styled_df_Labor


,1,2,3,4,year
Unit to be Produced,1060.000000,1260.000000,1600.000000,1800.000000,5720.000000
Direct Labour Hour,0.120000,0.120000,0.120000,0.120000,0.120000
Total Hour Needs,127.200000,151.200000,192.000000,216.000000,686.400000
Wage Labor Hour,10.000000,10.000000,10.000000,10.000000,10.000000
Total Direct Labor Cost,1272.000000,1512.000000,1920.000000,2160.000000,6864.000000


# **Total Overhead Budget**

# Manufacturing Overhead Budget



### Business Scenario

Following the completion of the **Direct Labor Budget**, **Texas Rex Apparel** must estimate its **Manufacturing Overhead Costs** for the upcoming fiscal year.

Manufacturing overhead consists of all indirect production costs that cannot be directly traced to individual T-shirts but are necessary to support the manufacturing process. These costs include both **variable** and **fixed** manufacturing overhead.

Management has established the following manufacturing overhead cost assumptions:

### Variable Manufacturing Overhead

- **$5.00 per direct labor hour**

### Fixed Manufacturing Overhead

- **$1,645 per quarter**

The fixed manufacturing overhead includes **$540 of depreciation expense** each quarter. Since depreciation is a **non-cash expense**, it should be identified separately for use in the **Cash Budget**.

The Manufacturing Overhead Budget will estimate total production overhead costs for each quarter and provide the information required to calculate product manufacturing costs and prepare subsequent budgets.

---

## Required

Using the **Direct Labor Budget** and the manufacturing overhead assumptions provided above:

1. Calculate the **variable manufacturing overhead** for each quarter.

2. Calculate the **total manufacturing overhead**, including both variable and fixed overhead costs.

3. Identify the **depreciation expense** included in manufacturing overhead for each quarter.

4. Prepare the **Manufacturing Overhead Budget** for each quarter and for the fiscal year.

In [ ]:
# --- OVERHEAD BUDGET ---

# Source Data
Variable_Overhead_Cost = Budgeting.loc[Budgeting['Account'] == 'Variable overhead cost rate'].iloc[:, 1:]
Fixed_Overhead_Cost = Budgeting.loc[Budgeting['Account'] == 'Budgeted Fixed overhead'].iloc[:, 1:]

# Calculation
# PERBAIKAN: Menggunakan Total_Hour_Needs dari perhitungan Direct Labor sebelumnya
Budgeted_Variable_Cost = Total_Hour_Needs * Variable_Overhead_Cost.values
Total_Overhead_Cost = Budgeted_Variable_Cost + Fixed_Overhead_Cost.values

# DataFrame
# Perhatikan kita memanggil df_Total_Hour_Needs yang sudah di-assign sebelumnya
df_Variable_Overhead_Cost = pd.DataFrame(Variable_Overhead_Cost.values, columns=Quarter, index=['Variable Overhead Cost Rate'])
df_Budgeted_Variable_Cost = pd.DataFrame(Budgeted_Variable_Cost, columns=Quarter, index=['Budgeted Variable Overhead'])
df_Fixed_Overhead_Cost = pd.DataFrame(Fixed_Overhead_Cost.values, columns=Quarter, index=['Budgeted Fixed Overhead'])
df_Total_Overhead_Cost = pd.DataFrame(Total_Overhead_Cost, columns=Quarter, index=['Total Overhead Cost'])

# Table Overhead Budget
# Menggabungkan df_Total_Hour_Needs di baris pertama agar visual outputnya persis seperti di Excel
df_Overhead_Budget = pd.concat([
    df_Total_Hour_Needs,
    df_Variable_Overhead_Cost,
    df_Budgeted_Variable_Cost,
    df_Fixed_Overhead_Cost,
    df_Total_Overhead_Cost
])

# Year Columns
# Logic tetap sama: Untuk rate kita ambil rata-rata (mean), sisanya dijumlahkan (sum)
df_Overhead_Budget['year'] = df_Overhead_Budget.apply(
    lambda row: row.mean() if row.name in ['Variable Overhead Cost Rate'] else row.sum(),
    axis=1
)

# Title
styled_df_Overhead = (
    df_Overhead_Budget.style
    .set_caption("Overhead Budget")
)

styled_df_Overhead

,1,2,3,4,year
Total Hour Needs,127.200000,151.200000,192.000000,216.000000,686.400000
Variable Overhead Cost Rate,5.000000,5.000000,5.000000,5.000000,5.000000
Budgeted Variable Overhead,636.000000,756.000000,960.000000,1080.000000,3432.000000
Budgeted Fixed Overhead,1645.000000,1645.000000,1645.000000,1645.000000,6580.000000
Total Overhead Cost,2281.000000,2401.000000,2605.000000,2725.000000,10012.000000


# **Ending Inventory**


As the financial analyst of **Texas Rex Apparel**, you have completed the Production Budget, Direct Materials Purchases Budget, Direct Labor Budget, and Manufacturing Overhead Budget for the upcoming fiscal year.

Management now requires an estimate of the **unit manufacturing cost** of each finished T-shirt. This information will be used to determine the value of ending finished goods inventory and to support financial reporting and budgeting decisions.

The unit product cost should include all manufacturing cost components:

- Direct Materials
- Direct Labor
- Manufacturing Overhead

The calculations should be based on the budgets prepared in the previous sections.

---

## Required

1. Calculate the **unit product cost** of one finished T-shirt.

2. Prepare the **Ending Finished Goods Inventory Budget** for each quarter and for the year.

In [ ]:
#Ending_Inventory_Budget
Ending_Inventory_Q4 = Budgeting.loc[Budgeting['Account'] == 'Ending Inventory'].iloc[:,-1]
Unit_Cost = 6.95
#Calculation
Total_Ending_inventory = Ending_Inventory_Q4.values *Unit_Cost

#DataFrame
Quarter_4 = Budgeting.columns[-1:]
df_Total_Ending_inventory = pd.DataFrame(Total_Ending_inventory, columns = Quarter_4, index = ['Total Ending Inventory'])

# Title
styled_df_Total_Ending_inventory = (
    df_Total_Ending_inventory.style
    .set_caption("Ending Inventory Budget")
)

styled_df_Total_Ending_inventory



,4
Total Ending Inventory,1390.000000


# **Perhitungan COGS**

After completing the **Production Budget**, **Direct Materials Purchases Budget**, **Direct Labor Budget**, **Manufacturing Overhead Budget**, and **Ending Finished Goods Inventory Budget**, management of **Texas Rex Apparel** now requires a **Cost of Goods Sold (COGS) Budget** for the upcoming fiscal year.

The COGS Budget will be used to estimate the total manufacturing costs associated with goods sold during the year. This budget is essential for preparing the Budgeted Income Statement and evaluating the company's expected gross profit.

The company reports a **beginning finished goods inventory** valued at **$1,251**.

The Cost of Goods Sold Budget should be prepared using information from the previously completed operating budgets.

---

## Required

Prepare the **Cost of Goods Sold (COGS) Budget** for each quarter and for the year

In [ ]:
# --- DIRECT MATERIAL USED BUDGET ---

# 1. Beginning Raw Material (Dari dataset)
# Pastikan nama akun sesuai dengan yang ada di csv ('Raw Material ')
Beginning_Raw_Material = Budgeting.loc[Budgeting['Account'] == 'Raw Material '].iloc[:,1].values[0]

# 2. Total Direct Material Purchased (Menggunakan perhitunganmu sebelumnya)
Total_Direct_Material_Purchased = df_Total_Direct_Material.loc['Total Direct Material', 'year']

# 3. Ending Direct Material (Persediaan Q4 dikali harga masing-masing bahan)
Ending_DM_A_Q4 = Budgeting.loc[Budgeting['Account'] == 'Ending Inventory Direct Material A'].iloc[:,-1].values[0]
Ending_DM_B_Q4 = Budgeting.loc[Budgeting['Account'] == 'Ending Inventory Direct Material B'].iloc[:,-1].values[0]
Price_DM_A = Budgeting.loc[Budgeting['Account'] == 'Direct material A cost (finish Good)'].iloc[:,1].values[0]
Price_DM_B = Budgeting.loc[Budgeting['Account'] == 'Direct material B cost (finish Good)'].iloc[:,1].values[0]

Ending_Direct_Material = (Ending_DM_A_Q4 * Price_DM_A) + (Ending_DM_B_Q4 * Price_DM_B)

# 4. Kalkulasi Direct Material Used
Direct_Material_Used = Beginning_Raw_Material + Total_Direct_Material_Purchased - Ending_Direct_Material

# DataFrame
year_col = ['year']
df_Direct_Material_Used = pd.DataFrame(
    [Beginning_Raw_Material, Total_Direct_Material_Purchased, Ending_Direct_Material, Direct_Material_Used],
    columns=year_col,
    index=['Beginning Raw Material', 'Direct Material Purchased', 'Ending Direct Material', 'Direct Material Used']
)

# Title
styled_df_DM_Used = (
    df_Direct_Material_Used.style
    .set_caption("Direct Material Used Budget")
)
styled_df_DM_Used

,year
Beginning Raw Material,252.000000
Direct Material Purchased,23052.000000
Ending Direct Material,424.000000
Direct Material Used,22880.000000


In [ ]:
# --- BUDGETED COST OF GOODS SOLD ---

# 1. Ekstrak nilai murni (float/scalar)
Direct_Labor_Used = float(df_Direct_Labor_Budget.loc['Total Direct Labor Cost', 'year'])
Overhead_Used = float(df_Overhead_Budget.loc['Total Overhead Cost', 'year'])
Total_Ending_Inventory_Val = float(df_Total_Ending_inventory.iloc[0, 0])

# 2. PERBAIKAN: Tarik Beginning Finished Goods dari dataset secara dinamis
# Menggunakan str.contains agar aman dari salah ketik atau extra spasi (' ') di Excel
Beginning_Finished_Goods = float(
    Budgeting.loc[Budgeting['Account'].str.contains('Finished Good', case=False, na=False)].iloc[:, 1].values[0]
)

# 3. Kalkulasi Akuntansi
Budgeted_Manufacturing_Cost = Direct_Material_Used + Direct_Labor_Used + Overhead_Used
Cost_of_Goods_Available_for_Sales = Budgeted_Manufacturing_Cost + Beginning_Finished_Goods
Cost_of_Good_Sold = Cost_of_Goods_Available_for_Sales - Total_Ending_Inventory_Val

# 4. Pembuatan DataFrame COGS
df_Budgeted_Cost_of_Good_Sold = pd.DataFrame(
    [
        Direct_Material_Used,
        Direct_Labor_Used,
        Overhead_Used,
        Budgeted_Manufacturing_Cost,
        Beginning_Finished_Goods,
        Cost_of_Goods_Available_for_Sales,
        Total_Ending_Inventory_Val,
        Cost_of_Good_Sold
    ],
    columns=['year'],
    index=[
        'Direct Material Used',
        'Direct Labor Used',
        'Overhead',
        'Budgeted Manufacturing Costs',
        'Beginning Finished Goods',
        'Cost of Goods Available for Sales',
        'Less : Ending Finished Goods',
        'Cost of Good Sold'
    ]
)

# 5. Visualisasi Tabel
styled_df_Budgeted_Cost_of_Good_Sold = (
    df_Budgeted_Cost_of_Good_Sold.style
    .set_caption("Budgeted Cost of Goods Sold")
)
styled_df_Budgeted_Cost_of_Good_Sold

,year
Direct Material Used,22880.000000
Direct Labor Used,6864.000000
Overhead,10012.000000
Budgeted Manufacturing Costs,39756.000000
Beginning Finished Goods,1251.000000
Cost of Goods Available for Sales,41007.000000
Less : Ending Finished Goods,1390.000000
Cost of Good Sold,39617.000000


# **BUDGETED SALES & ADMINISTRATIVE**

As part of the annual budgeting process, **Texas Rex Apparel** must estimate its **Selling and Administrative (S&A) Expenses** for the upcoming fiscal year.

These costs include both **variable selling expenses**, which fluctuate with sales volume, and **fixed operating expenses**, which remain constant regardless of sales activity.

The company has established the following cost assumptions:

### Variable Selling Expense

- **$0.10 per unit sold**

### Fixed Selling and Administrative Expenses

- **Salaries:** $1,420 per quarter
- **Utilities:** $50 per quarter
- **Depreciation:** $150 per quarter
- **Advertising Expense:**
  - Q1: $100
  - Q2: $200
  - Q3: $800
  - Q4: $500

The Selling and Administrative Expense Budget will provide the operating expense estimates required for preparing the **Budgeted Income Statement**.

---

## Required

Prepare the **Selling and Administrative Expense Budget** for each quarter and for the year.

In [ ]:
# --- BUDGETED SALES & ADMINISTRATIVE ---

# Source Data
Variable_Cost_SA = Budgeting.loc[Budgeting['Account'] == 'Variable cost S&A'].iloc[:,1:]
Salaries_SA = Budgeting.loc[Budgeting['Account'] == 'Salaries S&A'].iloc[:,1:]
Utilities_SA = Budgeting.loc[Budgeting['Account'] == 'Utilities S&A'].iloc[:,1:]
Advertising_SA = Budgeting.loc[Budgeting['Account'] == 'Advertising S&A'].iloc[:,1:]
Depreciation_SA = Budgeting.loc[Budgeting['Account'] == 'Depreciation S&A'].iloc[:,1:]

# Calculation
Total_Variable_Cost_SA = Sales_Unit.values * Variable_Cost_SA.values
Total_Fixed_Cost_SA = (
    Salaries_SA.values
    + Utilities_SA.values
    + Advertising_SA.values
    + Depreciation_SA.values
)
Total_SA_Cost = Total_Variable_Cost_SA + Total_Fixed_Cost_SA

# DataFrames
df_Variable_Cost_SA = pd.DataFrame(Variable_Cost_SA.values, columns=Quarter, index=['Variable Cost S&A'])
df_Total_Variable_Cost_SA = pd.DataFrame(Total_Variable_Cost_SA, columns=Quarter, index=['Total Variable Cost S&A'])

df_Salaries_SA = pd.DataFrame(Salaries_SA.values, columns=Quarter, index=['Salaries S&A'])
df_Utilities_SA = pd.DataFrame(Utilities_SA.values, columns=Quarter, index=['Utilities S&A'])
df_Advertising_SA = pd.DataFrame(Advertising_SA.values, columns=Quarter, index=['Advertising S&A'])
df_Depreciation_SA = pd.DataFrame(Depreciation_SA.values, columns=Quarter, index=['Depreciation S&A'])

df_Total_Fixed_Cost_SA = pd.DataFrame(Total_Fixed_Cost_SA, columns=Quarter, index=['Total Fixed Cost S&A'])
df_Total_SA_Cost = pd.DataFrame(Total_SA_Cost, columns=Quarter, index=['Total S&A Cost'])

# Table
df_Budgeted_SA = pd.concat([
    df_Sales_Unit,
    df_Variable_Cost_SA,
    df_Total_Variable_Cost_SA,
    df_Salaries_SA,
    df_Utilities_SA,
    df_Advertising_SA,
    df_Depreciation_SA,
    df_Total_Fixed_Cost_SA,
    df_Total_SA_Cost
])

# Year column
df_Budgeted_SA['year'] = df_Budgeted_SA.apply(
    lambda row: row.mean() if row.name in ['Variable Cost S&A'] else row.sum(),
    axis=1
)

# Title
styled_df_Budgeted_SA = (
    df_Budgeted_SA.style
    .set_caption("Sales & Administrative Budget")
)

styled_df_Budgeted_SA

,1,2,3,4,year
Unit,1000.000000,1200.000000,1500.000000,2000.000000,5700.000000
Variable Cost S&A,0.100000,0.100000,0.100000,0.100000,0.100000
Total Variable Cost S&A,100.000000,120.000000,150.000000,200.000000,570.000000
Salaries S&A,1420.000000,1420.000000,1420.000000,1420.000000,5680.000000
Utilities S&A,50.000000,50.000000,50.000000,50.000000,200.000000
Advertising S&A,100.000000,200.000000,800.000000,500.000000,1600.000000
Depreciation S&A,150.000000,150.000000,150.000000,150.000000,600.000000
Total Fixed Cost S&A,1720.000000,1820.000000,2420.000000,2120.000000,8080.000000
Total S&A Cost,1820.000000,1940.000000,2570.000000,2320.000000,8650.000000


# **Income Statement**

After completing the **Sales Budget**, **Cost of Goods Sold (COGS) Budget**, **Selling and Administrative Expense Budget**, and **Cash Budget**, management of **Texas Rex Apparel** is ready to prepare the company's **Budgeted Income Statement** for the upcoming fiscal year.

The Budgeted Income Statement summarizes the company's expected financial performance by combining projected revenues, manufacturing costs, operating expenses, and income taxes. This report serves as a key financial planning tool for evaluating expected profitability and supporting strategic decision-making.

For budgeting purposes, the company estimates that **income tax expense will be calculated using a tax rate of 40%**.

The Budgeted Income Statement should be prepared using information from the previously completed operating budgets.

---

## Required

Prepare the **Budgeted Income Statement** for the fiscal year.

In [ ]:
# --- BUDGETED INCOME STATEMENT ---

# 1. Pastikan df_Sales_Budget memiliki kolom 'year'
if 'year' not in df_Sales_Budget.columns:
    df_Sales_Budget['year'] = df_Sales_Budget.apply(lambda row : row.mean() if row.name == 'Price Sell' else row.sum(), axis = 1)

# 2. Ekstrak nilai total tahunan sebagai skalar murni (float)
Sales_Year = float(df_Sales_Budget.loc['Budgeted Sales', 'year'])
COGS_Year = float(df_Budgeted_Cost_of_Good_Sold.loc['Cost of Good Sold', 'year'])
S_A_Year = float(df_Budgeted_SA.loc['Total S&A Cost', 'year'])
Interest_Expense = float(Budgeting.loc[Budgeting['Account'] == 'Interest Expense'].iloc[:, 1].values[0])

# 3. Kalkulasi Akuntansi
Gross_Margin = Sales_Year - COGS_Year
Operating_Income = Gross_Margin - S_A_Year
Income_Before_Income_Taxes = Operating_Income - Interest_Expense
Income_Taxes = Income_Before_Income_Taxes * 0.4
Net_Income = Income_Before_Income_Taxes - Income_Taxes

# 4. Pembuatan DataFrame Laba Rugi dengan Pembulatan (.round)
year_col = df_Direct_Material_A_Budget.columns[-1:]

df_Income_Statement_Budget = pd.DataFrame(
    [
        Sales_Year,
        COGS_Year,
        Gross_Margin,
        S_A_Year,
        Operating_Income,
        Interest_Expense,
        Income_Before_Income_Taxes,
        Income_Taxes,
        Net_Income
    ],
    columns=year_col,
    index=[
        'Sales',
        'Less : Cost of Good Sold',
        'Gross Margin',
        'Less : Selling & Administrative Expenses',
        'Operating Income',
        'Less: Interest Expenses',
        'Income before income taxes',
        'Less: Income taxes',
        'Net Income'
    ]
).round(0) # Membulatkan keseluruhan nilai di dalam tabel menjadi bilangan bulat

# 5. Visualisasi Tabel dengan Format Akuntansi
styled_df_Income_Statement = (
    df_Income_Statement_Budget.style
    .format("{:,.0f}") # Menambahkan pemisah ribuan (contoh: 5,954)
    .set_caption("Budgeted Income Statement")
)

styled_df_Income_Statement

,year
Sales,"57,000"
Less : Cost of Good Sold,"39,617"
Gross Margin,"17,383"
Less : Selling & Administrative Expenses,"8,650"
Operating Income,"8,733"
Less: Interest Expenses,60
Income before income taxes,"8,673"
Less: Income taxes,"3,469"
Net Income,"5,204"


# **CASH RECEIVE (COLLECTIONS) BUDGET**

To support cash flow planning, **Texas Rex Apparel** prepares a **Cash Collections Budget** to estimate the timing of cash receipts from customer sales.

Based on the company's historical collection patterns:

- **25%** of total sales are collected **in cash at the time of sale**.
- The remaining **75%** are made on **credit**.
- Of the credit sales:
  - **90%** are collected in the **following quarter**.
  - **10%** are collected **two quarters after the sale**.

The projected sales revenue for the upcoming fiscal year is as follows:

| Quarter | Budgeted Sales ($) |
|:--------|-------------------:|
| Q1 | 10,000 |
| Q2 | 12,000 |
| Q3 | 15,000 |
| Q4 | 20,000 |

At the beginning of the budget year, the company has an **Accounts Receivable balance of $1,350**, representing outstanding credit sales from the previous year. Management expects this entire balance to be collected during the **first quarter**.

The Cash Collections Budget will be used as the primary input for the company's **Cash Budget**.

---

## Required

1. Calculate the expected **cash collections** for each quarter of the budget year.

2. Prepare a **Schedule of Cash Collections from Sales** for each quarter and for the year.

In [ ]:
# --- CASH RECEIVE (COLLECTIONS) BUDGET ---

# 1. Asumsi Persentase Penagihan (Sesuai Prototype Excel)
pct_cash_sales = 0.25
pct_credit_sales = 0.75
pct_collected_current = 0.90
pct_collected_next = 0.10

# 2. Tarik Data Budgeted Sales untuk Q1-Q4
# Menggunakan iloc[0:4] untuk menghindari terambilnya kolom 'year' jika ada
Sales_Vals = df_Budgeted_Sales.loc['Budgeted Sales'].iloc[0:4].values.astype(float)

# 3. Tarik Data Account Receivable (Piutang Q4 Tahun Lalu) dari dataset secara dinamis
Prior_Year_AR = float(
    Budgeting.loc[Budgeting['Account'].str.contains('Account Receivable', case=False, na=False)].iloc[:, 1].values[0]
)

# 4. Hitung Dasar Penjualan Tunai & Kredit
Cash_Sales = Sales_Vals * pct_cash_sales
Credit_Sales = Sales_Vals * pct_credit_sales

# 5. Susun Matriks Penagihan (Cash Inflow per Kuartal)
Rec_Prior_Q4 = [Prior_Year_AR, 0.0, 0.0, 0.0]
Rec_Q1       = [Credit_Sales[0] * pct_collected_current, Credit_Sales[0] * pct_collected_next, 0.0, 0.0]
Rec_Q2       = [0.0, Credit_Sales[1] * pct_collected_current, Credit_Sales[1] * pct_collected_next, 0.0]
Rec_Q3       = [0.0, 0.0, Credit_Sales[2] * pct_collected_current, Credit_Sales[2] * pct_collected_next]
Rec_Q4       = [0.0, 0.0, 0.0, Credit_Sales[3] * pct_collected_current] # Sisa 10% akan masuk ke neraca (Ending AR)

# 6. Kalkulasi Total Penerimaan Kas
Total_Receipt_Cash = [
    Cash_Sales[i] + Rec_Prior_Q4[i] + Rec_Q1[i] + Rec_Q2[i] + Rec_Q3[i] + Rec_Q4[i]
    for i in range(4)
]

# 7. Pembuatan DataFrame Cash Receive
Quarter_Cols = Budgeting.columns[1:5] # Memastikan nama kolom adalah 1, 2, 3, 4

df_Cash_Receive = pd.DataFrame(
    [
        Cash_Sales,
        Rec_Prior_Q4,
        Rec_Q1,
        Rec_Q2,
        Rec_Q3,
        Rec_Q4,
        Total_Receipt_Cash
    ],
    columns=Quarter_Cols,
    index=[
        'Cash Sales',
        'Quarter 4 (Prior year)',
        'Quarter 1',
        'Quarter 2',
        'Quarter 3',
        'Quarter 4',
        'Total Receipt Cash'
    ]
)

# 8. Visualisasi Tabel
styled_df_Cash_Receive = (
    df_Cash_Receive.style
    .format("{:,.0f}")
    .set_caption("Cash Receive Schedule")
)

styled_df_Cash_Receive

,1,2,3,4
Cash Sales,"2,500","3,000","3,750","5,000"
Quarter 4 (Prior year),"1,350",0,0,0
Quarter 1,"6,750",750,0,0
Quarter 2,0,"8,100",900,0
Quarter 3,0,0,"10,125","1,125"
Quarter 4,0,0,0,"13,500"
Total Receipt Cash,"10,600","11,850","14,775","19,625"


# **CASH FOR PAYMENT (DISBURSEMENTS FOR MATERIALS) BUDGET**

# Cash Disbursements for Materials Budget

## Case Study

As part of its annual budgeting process, **Texas Rex Apparel** purchases all direct materials on **credit** from its suppliers.

Based on the company's payment policy:

- **80%** of material purchases are paid in the **same quarter** in which the purchases are made.
- The remaining **20%** are paid in the **following quarter**.

At the beginning of the budget year, the company has an **Accounts Payable balance of $5,000**, representing unpaid material purchases made during the fourth quarter of the previous year. This balance will be settled during the **first quarter** of the current budget year.

Based on the **Direct Materials Purchases Budget**, the estimated material purchases for the year are:

| Quarter | Budgeted Material Purchases ($) |
|:--------|--------------------------------:|
| Q1 | 4,492 |
| Q2 | 5,176 |
| Q3 | 6,480 |
| Q4 | 6,904 |
The schedule of cash disbursements for materials will be used as an input to the company's **Cash Budget**.

---

## Required

Prepare a **Schedule of Cash Disbursements for Materials** for each quarter and for the year.

In [ ]:
# --- CASH FOR PAYMENT (DISBURSEMENTS FOR MATERIALS) BUDGET ---

# 1. Asumsi Pembayaran
pct_paid_current = 0.80
pct_paid_next = 0.20

# 2. Tarik Data Total Direct Material
# Mengambil data pembelian bahan baku (A+B) dari df_Total_Direct_Material
Purchases_Vals = df_Total_Direct_Material.loc['Total Direct Material'].iloc[0:4].values.astype(float)

# 3. Tarik Hutang Usaha (Account Payable) Q4 Tahun Lalu dari dataset
Prior_Year_AP = float(
    Budgeting.loc[Budgeting['Account'].str.contains('Account Payable', case=False, na=False)].iloc[:, 1].values[0]
)

# 4. Susun Matriks Pembayaran (Cash Outflow per Kuartal)
Pay_Prior_Q4 = [Prior_Year_AP, 0.0, 0.0, 0.0]
Pay_Q1       = [Purchases_Vals[0] * pct_paid_current, Purchases_Vals[0] * pct_paid_next, 0.0, 0.0]
Pay_Q2       = [0.0, Purchases_Vals[1] * pct_paid_current, Purchases_Vals[1] * pct_paid_next, 0.0]
Pay_Q3       = [0.0, 0.0, Purchases_Vals[2] * pct_paid_current, Purchases_Vals[2] * pct_paid_next]
Pay_Q4       = [0.0, 0.0, 0.0, Purchases_Vals[3] * pct_paid_current] # Sisa 20% akan masuk ke Neraca sebagai Ending AP

# 5. Kalkulasi Total Pengeluaran Kas
Total_Cash_Needed = [
    Pay_Prior_Q4[i] + Pay_Q1[i] + Pay_Q2[i] + Pay_Q3[i] + Pay_Q4[i]
    for i in range(4)
]

# 6. Pembuatan DataFrame Cash for Payment
Quarter_Cols = Budgeting.columns[1:5]

df_Cash_Payment = pd.DataFrame(
    [
        Pay_Prior_Q4,
        Pay_Q1,
        Pay_Q2,
        Pay_Q3,
        Pay_Q4,
        Total_Cash_Needed
    ],
    columns=Quarter_Cols,
    index=[
        'Quarter 4 (Prior year)',
        'Quarter 1',
        'Quarter 2',
        'Quarter 3',
        'Quarter 4',
        'Total Cash Needed'
    ]
)

# 7. Visualisasi Tabel
styled_df_Cash_Payment = (
    df_Cash_Payment.style
    .format("{:,.1f}") # Format 1 desimal untuk menyembunyikan angka di belakang koma yang terlalu panjang (float imprecision)
    .set_caption("Cash for Payment Budget")
)

styled_df_Cash_Payment

,1,2,3,4
Quarter 4 (Prior year),"1,000.0",0.0,0.0,0.0
Quarter 1,"3,593.6",898.4,0.0,0.0
Quarter 2,0.0,"4,140.8","1,035.2",0.0
Quarter 3,0.0,0.0,"5,184.0","1,296.0"
Quarter 4,0.0,0.0,0.0,"5,523.2"
Total Cash Needed,"4,593.6","5,039.2","6,219.2","6,819.2"


# **CASH BUDGET**


As the final step in the operating budgeting process, **Texas Rex Apparel** is preparing its **Cash Budget** for the upcoming fiscal year. The Cash Budget will help management evaluate the company's liquidity position, determine financing requirements, and ensure that sufficient cash is available to support daily operations.

The Cash Budget should incorporate all projected cash inflows and outflows developed in the previous budgeting schedules, including cash collections from customers and cash payments for operating activities.

The company has established the following financial policies and assumptions:

### Financing Policy

- The company requires a **minimum cash balance of $1,000** at the end of each quarter.
- If the ending cash balance falls below this minimum, the company may borrow funds in **multiples of $1,000**.
- Borrowings are made **at the beginning of each quarter**.
- Loan repayments, including interest, are made **at the end of each quarter**.
- The annual interest rate on borrowings is **12%**, calculated only on the outstanding principal.

### Non-Cash Expenses

The following depreciation expenses are included in the operating budgets but **do not require cash payments**:

- Manufacturing Overhead Depreciation: **$540 per quarter**
- Selling and Administrative Depreciation: **$150 per quarter**

### Capital Expenditure

To expand production capacity, the company plans to purchase additional **screen-printing equipment** during the **first quarter**.

- Equipment Cost: **$6,500**
- The purchase will be paid **entirely in cash** during Q1.
- Management expects to finance this investment through a combination of operating cash flows and short-term borrowings if necessary.

### Income Taxes

- Estimated income tax payable for the year: **$3,469**
- The tax payment will be made **at the end of the fourth quarter**.

### Beginning Cash Balance

- Cash available at the beginning of the budget year: **$5,200**

### Additional Assumption

- All amounts presented in the budget are **rounded to the nearest dollar**.

---

## Required

Prepare the **Cash Budget** for each quarter and for the fiscal year.

In [ ]:
import numpy as np
import pandas as pd

# --- CASH BUDGET (ANTI-KEYERROR VERSION) ---

# 1. Ekstrak Data Aliran Kas Masuk (Inflows)
# Menggunakan .iloc[-1] untuk memaksa Python mengambil baris PALING BAWAH (Total)
# Ini membuat kode kebal terhadap salah ketik atau spasi tersembunyi di nama indeks
collections = df_Cash_Receive.iloc[-1, 0:4].values.flatten().astype(float)

# Tarik depresiasi dari dataset mentah
depr_oh = Budgeting.loc[Budgeting['Account'].str.contains('Depreciatin Overhead', case=False, na=False)].iloc[:, 1:5].values.flatten().astype(float)
depr_sa = Budgeting.loc[Budgeting['Account'].str.contains('Depreciation S&A', case=False, na=False)].iloc[:, 1:5].values.flatten().astype(float)

# 2. Ekstrak Data Aliran Kas Keluar (Disbursements)
disb_rm = np.array(Total_Cash_Needed, dtype=float)
disb_dl = df_Direct_Labor_Budget.loc['Total Direct Labor Cost'].iloc[0:4].values.astype(float)
disb_oh = df_Overhead_Budget.loc['Total Overhead Cost'].iloc[0:4].values.astype(float)
disb_sa = df_Budgeted_SA.loc['Total S&A Cost'].iloc[0:4].values.astype(float)

# PERBAIKAN: Menggunakan nama baris yang spesifik dan presisi
tax_val = float(df_Income_Statement_Budget.loc['Less: Income taxes'].iloc[-1])
taxes = np.array([0.0, 0.0, 0.0, tax_val])

# Tarik pengeluaran modal (Equipment) secara dinamis
equip_val = float(Budgeting.loc[Budgeting['Account'].str.contains('New Building', case=False, na=False)].iloc[:, 1].values[0])
equip = np.array([equip_val, 0.0, 0.0, 0.0])

# 3. Ekstrak Data Financing
borrow_val = float(Budgeting.loc[Budgeting['Account'].str.contains('Borrowing', case=False, na=False)].iloc[:, 1].values[0])
borrowing = np.array([borrow_val, 0.0, 0.0, 0.0])

repayment = np.array([0.0, -borrow_val, 0.0, 0.0])

interest_val = float(Budgeting.loc[Budgeting['Account'].str.contains('Interest Expense', case=False, na=False)].iloc[:, 1].values[0])
interest = np.array([0.0, -interest_val, 0.0, 0.0])

total_financing = borrowing + repayment + interest

# 4. Inisiasi Array untuk Perhitungan Iteratif
beg_cash = np.zeros(4)
beg_cash[0] = 5200.0  # Saldo awal kas Q1

tot_cash_avail = np.zeros(4)
tot_disb = np.zeros(4)
excess_cash = np.zeros(4)
end_cash = np.zeros(4)

# 5. Iterasi Kalkulasi per Kuartal (Cascading Cash Balance)
for i in range(4):
    tot_cash_avail[i] = beg_cash[i] + collections[i] + depr_oh[i] + depr_sa[i]
    tot_disb[i] = -(disb_rm[i] + disb_dl[i] + disb_oh[i] + disb_sa[i] + taxes[i] + equip[i])
    excess_cash[i] = tot_cash_avail[i] + tot_disb[i]
    end_cash[i] = excess_cash[i] + total_financing[i]

    # Passing saldo akhir ke saldo awal kuartal berikutnya
    if i < 3:
        beg_cash[i+1] = end_cash[i]

# 6. Pembuatan DataFrame Cash Budget
Quarter_Cols = ['1', '2', '3', '4']

df_Cash_Budget = pd.DataFrame(
    [
        beg_cash, collections, depr_oh, depr_sa, tot_cash_avail,
        -disb_rm, -disb_dl, -disb_oh, -disb_sa, -taxes, -equip, tot_disb,
        excess_cash, borrowing, repayment, interest, total_financing, end_cash
    ],
    columns=Quarter_Cols,
    index=[
        'Beginning Cash Balance', 'Cash Sales and Collection on Account', 'Depreciation Overhead', 'Depreciation S&A', 'Total Cash available',
        'Raw Materials', 'Direct labor', 'Overhead', 'Selling and Administrative expense', 'Income Taxes', 'Equipment', 'Total Disbursment',
        'Excess /Deficiency Cash available over needs', 'Borrowing', 'Repayment', 'Interest', 'Total Financing', 'Ending Cash Balance'
    ]
)

# 7. Penyesuaian Kolom 'year' untuk Saldo Kumulatif Akuntansi
df_Cash_Budget['year'] = df_Cash_Budget.apply(
    lambda row:
        row.iloc[0] if row.name == 'Beginning Cash Balance' else
        row.iloc[3] if row.name == 'Ending Cash Balance' else
        row.sum(),
    axis=1
)

# --- PERBAIKAN: Membulatkan seluruh angka ke bilangan bulat ---
df_Cash_Budget = df_Cash_Budget.round(0)

# 8. Visualisasi Tabel dengan Formatting Akuntansi
styled_df_Cash_Budget = (
    df_Cash_Budget.style
    .format("{:,.0f}") # Memastikan pemisah ribuan tampil tanpa angka desimal
    .set_caption("Cash Budget")
)

styled_df_Cash_Budget

,1,2,3,4,year
Beginning Cash Balance,"5,200","1,023","1,611","3,762","5,200"
Cash Sales and Collection on Account,"10,600","11,850","14,775","19,625","56,850"
Depreciation Overhead,540,540,540,540,"2,160"
Depreciation S&A,150,150,150,150,600
Total Cash available,"16,490","13,563","17,076","24,077","71,207"
Raw Materials,"-4,594","-5,039","-6,219","-6,819","-22,671"
Direct labor,"-1,272","-1,512","-1,920","-2,160","-6,864"
Overhead,"-2,281","-2,401","-2,605","-2,725","-10,012"
Selling and Administrative expense,"-1,820","-1,940","-2,570","-2,320","-8,650"
Income Taxes,-0,-0,-0,"-3,469","-3,469"


# **Budgeted Statement of Financial Performance**

# Budgeted Statement of Financial Position

## Case Study

After completing the **Operating Budgets**, **Budgeted Income Statement**, and **Cash Budget**, **Texas Rex Apparel** is ready to prepare its **Budgeted Statement of Financial Position (SOFP)** as of **December 31**.

The Budgeted Statement of Financial Position summarizes the company's expected financial position at the end of the fiscal year by combining projected asset, liability, and equity balances. It serves as a key financial statement for evaluating liquidity, solvency, and the overall financial health of the business.

The ending balances should be determined using information obtained from the previously completed budgeting schedules, including:

- Cash Budget
- Direct Materials Purchases Budget
- Ending Finished Goods Inventory Budget
- Capital Expenditure Budget
- Budgeted Income Statement

The company should also incorporate the effects of depreciation, financing activities, and retained earnings generated during the budget period.

The completed Budgeted Statement of Financial Position must satisfy the fundamental accounting equation:

> **Assets = Liabilities + Owner's Equity**

---

## Required

Prepare the **Budgeted Statement of Financial Position (SOFP)** as of **December 31** using the results from the previously completed budgets.

In [ ]:
import numpy as np
import pandas as pd

# --- BUDGETED STATEMENT OF FINANCIAL POSITION (SOFP) ---

# 1. Ekstrak Data Current Assets
cash_end = float(df_Cash_Budget.loc['Ending Cash Balance', 'year'])
ar_end = float(Sales_Vals[3] * pct_credit_sales * pct_collected_next) # 10% dari Credit Sales Q4 yang belum tertagih
rm_end = float(Ending_Direct_Material) # Tarik dari modul Direct Material Used
fg_end = float(Total_Ending_Inventory_Val) # Tarik dari modul Ending Inventory / COGS

total_current_assets = cash_end + ar_end + rm_end + fg_end

# 2. Ekstrak Data Non-Current Assets
land = float(Budgeting.loc[Budgeting['Account'].str.contains('Land', case=False, na=False)].iloc[:, 1].values[0])

# Base building ditambah dengan equipment baru yang dibeli di Q1 (6500)
base_building = float(Budgeting.loc[Budgeting['Account'] == 'Building and equipment '].iloc[:, 1].values[0])
building = base_building + equip_val

# Akumulasi depresiasi awal DITAMBAH (secara negatif) dengan beban depresiasi tahun berjalan
base_depr = float(Budgeting.loc[Budgeting['Account'].str.contains('accumulated', case=False, na=False)].iloc[:, 1].values[0])
acc_depr = base_depr - (sum(depr_oh) + sum(depr_sa))

total_non_current_assets = land + building + acc_depr
total_assets = total_current_assets + total_non_current_assets

# 3. Ekstrak Data Liabilities
ap_end = float(Purchases_Vals[3] * pct_paid_next) # 20% dari Pembelian Material Q4 yang belum dibayar
total_liabilities = ap_end

# 4. Ekstrak Data Equity
base_re = float(Budgeting.loc[Budgeting['Account'].str.contains('Retained Earning', case=False, na=False)].iloc[:, 1].values[0])
net_income = float(df_Income_Statement_Budget.loc['Net Income', 'year']) # Laba bersih masuk ke modal
re_end = base_re + net_income
total_equity = re_end

total_liabilities_and_equity = total_liabilities + total_equity

# 5. Susun DataFrame (Menyertakan np.nan untuk baris Header agar formatnya identik dengan Excel)
df_SOFP = pd.DataFrame(
    [
        np.nan, cash_end, ar_end, rm_end, fg_end, total_current_assets,
        np.nan, land, building, acc_depr, total_non_current_assets,
        total_assets,
        np.nan, np.nan, ap_end, total_liabilities,
        np.nan, re_end, total_equity, total_liabilities_and_equity
    ],
    columns=['Amount'],
    index=[
        'Current Asset',
        'Cash',
        'Account Receivable',
        'Raw Material Inventory',
        'Finished Good Inventory',
        'Total Current Asset',

        'Non Current Asset',
        'Land',
        'Building and equipment',
        'Accumulated Depreciation',
        'Total Non Current Asset',
        'Total Asset',

        'Liabilities & Equity',
        'Current Liabilities',
        'Account Payable',
        'Total Liabilities',

        'Owner Equity',
        'Retained Earnings',
        'Total Equity',
        'Total Liabilities & Equity'
    ]
)

# 6. Pembulatan (.round) dan Visualisasi
df_SOFP = df_SOFP.round(0)

styled_df_SOFP = (
    df_SOFP.style
    .format("{:,.0f}", na_rep="") # na_rep="" akan membuat baris header (np.nan) menjadi kosong secara visual
    .set_caption("Budgeted Statement of Financial Position")
)

styled_df_SOFP

,Amount
Current Asset,
Cash,"6,584"
Account Receivable,"1,500"
Raw Material Inventory,424
Finished Good Inventory,"1,390"
Total Current Asset,"9,898"
Non Current Asset,
Land,"1,100"
Building and equipment,"36,500"
Accumulated Depreciation,"-7,760"


# **Financial Statment Analysis**



After completing the **Master Budget** for the upcoming fiscal year, the management team of **Texas Rex Apparel** would like to evaluate the company's projected financial performance and financial position before implementing the budget.

As a Financial Planning & Analysis (FP&A) Analyst, your responsibility is not only to prepare the company's financial statements but also to interpret the results and identify potential financial strengths and risks.

Using the **Budgeted Income Statement** and **Budgeted Statement of Financial Position**, perform a comprehensive financial analysis to assess the company's expected profitability, liquidity, operating efficiency, and financial stability.

In addition to traditional financial ratio analysis, management also requests an assessment of the company's bankruptcy risk using the **Altman Z-Score** model. The results will support strategic planning, financing decisions, and overall business performance evaluation for the upcoming fiscal year.

---

## Required

Using the projected financial statements prepared in the previous sections:

1. Calculate the company's **Liquidity Ratios**.
2. Calculate the company's **Profitability Ratios**.
3. Calculate the company's **Efficiency (Activity) Ratios**.
4. Calculate the company's **Solvency (Leverage) Ratios**.
5. Calculate the **Altman Z-Score**.
6. Interpret the results and provide a brief financial performance assessment based on the calculated ratios.

In [ ]:
import pandas as pd

# --- FINANCIAL STATEMENT ANALYSIS ---

# 1. Ekstrak Data dari Budgeted Income Statement
sales = float(df_Income_Statement_Budget.loc['Sales', 'year'])
gross_margin = float(df_Income_Statement_Budget.loc['Gross Margin', 'year'])
op_income = float(df_Income_Statement_Budget.loc['Operating Income', 'year'])
net_income = float(df_Income_Statement_Budget.loc['Net Income', 'year'])

# 2. Ekstrak Data dari Statement of Financial Position (SOFP)
# Memanggil persis sesuai dengan nama index yang kita buat di df_SOFP
current_assets = float(df_SOFP.loc['Total Current Asset', 'Amount'])
cash = float(df_SOFP.loc['Cash', 'Amount'])
ar = float(df_SOFP.loc['Account Receivable', 'Amount'])

# Di kasus ini, seluruh kewajiban adalah Account Payable (Current Liability)
current_liabilities = float(df_SOFP.loc['Total Liabilities', 'Amount'])

total_assets = float(df_SOFP.loc['Total Asset', 'Amount'])
total_equity = float(df_SOFP.loc['Total Equity', 'Amount'])

# 3. Kalkulasi Rasio Keuangan (Key Performance Indicators)

# --- Liquidity Ratios ---
current_ratio = current_assets / current_liabilities
quick_ratio = (cash + ar) / current_liabilities

# --- Profitability Ratios ---
gross_profit_margin = gross_margin / sales
operating_profit_margin = op_income / sales
net_profit_margin = net_income / sales
roa = net_income / total_assets
roe = net_income / total_equity

# --- Efficiency Ratios ---
asset_turnover = sales / total_assets

# 4. Susun ke dalam DataFrame
df_Financial_Analysis = pd.DataFrame(
    [
        current_ratio,
        quick_ratio,
        gross_profit_margin,
        operating_profit_margin,
        net_profit_margin,
        roa,
        roe,
        asset_turnover
    ],
    columns=['Ratio / Margin'],
    index=[
        'Current Ratio (x)',
        'Quick Ratio (x)',
        'Gross Profit Margin (%)',
        'Operating Profit Margin (%)',
        'Net Profit Margin (%)',
        'Return on Assets / ROA (%)',
        'Return on Equity / ROE (%)',
        'Asset Turnover (x)'
    ]
)

# 5. Visualisasi Tabel dengan Formatting Dinamis
# Fungsi untuk memisahkan format persentase (%) dan format rasio kelipatan (x)
def format_ratios(val, index_name):
    if '%' in index_name:
        return f"{val:.2%}" # Format jadi persentase dengan 2 desimal (contoh: 15.50%)
    else:
        return f"{val:.2f}x" # Format jadi kali (contoh: 1.50x)

# Mengaplikasikan fungsi format ke DataFrame agar rapi
formatted_analysis = df_Financial_Analysis.copy()
formatted_analysis['Ratio / Margin'] = [
    format_ratios(val, idx) for val, idx in zip(df_Financial_Analysis['Ratio / Margin'], df_Financial_Analysis.index)
]

styled_df_Analysis = (
    formatted_analysis.style
    .set_caption("Financial Statement Analysis Dashboard")
)

styled_df_Analysis

,Ratio / Margin
Current Ratio (x),7.17x
Quick Ratio (x),5.85x
Gross Profit Margin (%),30.50%
Operating Profit Margin (%),15.32%
Net Profit Margin (%),9.13%
Return on Assets / ROA (%),13.10%
Return on Equity / ROE (%),13.57%
Asset Turnover (x),1.43x


In [ ]:
import pandas as pd

# ==========================================
# ALTMAN Z-SCORE (PREDICTIVE BANKRUPTCY MODEL)
# ==========================================

# 1. Menyiapkan Komponen Data dari Budget Sebelumnya
# Working Capital = Current Assets - Current Liabilities
working_capital = total_current_assets - current_liabilities
total_assets_z = total_assets
retained_earnings_z = re_end           # Diambil dari Saldo Laba Neraca
ebit_z = op_income                     # Laba Operasional (Operating Income)
book_value_equity_z = total_equity     # Sebagai proxy untuk Market Value of Equity
total_liabilities_z = total_liabilities
sales_z = sales

# 2. Menghitung 5 Rasio Pembentuk Z-Score (X1 - X5)
X1 = working_capital / total_assets_z
X2 = retained_earnings_z / total_assets_z
X3 = ebit_z / total_assets_z
X4 = book_value_equity_z / total_liabilities_z
X5 = sales_z / total_assets_z

# 3. Kalkulasi Altman Z-Score (Klasik / Original Model)
z_score = (1.2 * X1) + (1.4 * X2) + (3.3 * X3) + (0.6 * X4) + (1.0 * X5)

# 4. Klasifikasi Zona Kesehatan Finansial
if z_score > 2.99:
    zone = "Safe Zone (Finansial Sangat Sehat)"
elif 1.81 <= z_score <= 2.99:
    zone = "Grey Zone (Area Waspada)"
else:
    zone = "Distress Zone (Risiko Kebangkrutan Tinggi)"

# 5. Susun ke dalam DataFrame
df_Z_Score = pd.DataFrame(
    [X1, X2, X3, X4, X5, z_score],
    columns=['Result'],
    index=[
        'X1: Working Capital / Total Assets',
        'X2: Retained Earnings / Total Assets',
        'X3: EBIT / Total Assets',
        'X4: Book Value of Equity / Total Liabilities',
        'X5: Sales / Total Assets',
        'Altman Z-Score Final'
    ]
)

# 6. Visualisasi Tabel
styled_Z_Score = (
    df_Z_Score.style
    .format("{:.4f}") # Z-Score sangat sensitif, disarankan pakai 4 angka di belakang koma
    .set_caption(f"Altman Z-Score Analysis | Status: {zone}")
    .set_table_styles([{
        'selector': 'caption',
        'props': [('font-size', '14px'), ('font-weight', 'bold')]
    }])
)

display(styled_Z_Score)

,Result
X1: Working Capital / Total Assets,0.2143
X2: Retained Earnings / Total Assets,0.9652
X3: EBIT / Total Assets,0.2198
X4: Book Value of Equity / Total Liabilities,27.7788
X5: Sales / Total Assets,1.4344
Altman Z-Score Final,20.4355


# **Download dataset for Dashboard (Power BI)**

In [ ]:
import pandas as pd
import numpy as np
from IPython.display import display

# ====================================================================
# 1. PERSIAPAN MASTER DATASET (SINGLE SHEET WIDE FORMAT)
# ====================================================================
# Menggabungkan seluruh akun dari Sales Budget hingga Financial Analysis
# ke dalam satu format tabel mendatar (Wide Format) dengan baris sebagai Periode

periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY (Full Year)']

# --- Helper Functions untuk Padding Data (Menyesuaikan panjang array menjadi 5) ---
# Menjumlahkan Q1-Q4 untuk mendapatkan total FY (Cocok untuk Sales, Biaya, Cashflow)
def pad_yearly_sum(quarterly_array):
    return list(quarterly_array) + [sum(quarterly_array)]

# Mengambil nilai Q4 untuk mendapatkan total FY (Cocok untuk Ending Balance Kas / Piutang)
def pad_yearly_last(quarterly_array):
    return list(quarterly_array) + [quarterly_array[-1]]

# Membuat nilai Q1-Q4 kosong/NaN karena data hanya dihitung tahunan (Cocok untuk SOFP & Ratios)
def pad_annual_only(annual_value):
    return [np.nan, np.nan, np.nan, np.nan, annual_value]


# --- Ekstraksi Data Kuantitas (Volume) ---
# Menarik nilai 'Unit' penjualan dari df_Sales_Budget (Kolom Q1-Q4)
sales_units_q = df_Sales_Budget.loc['Unit'].iloc[0:4].values.flatten().astype(float)
# Catatan: Jika kamu ingin memasukkan Unit Produksi, kamu bisa un-comment baris di bawah ini:
# production_units_q = df_Production_Budget.iloc[-1, 0:4].values.flatten().astype(float)


# --- Ekstraksi Data Operasional Kuartalan ---
# Pastikan array ini ditarik dari variabel yang sudah sukses berjalan di atas
revenue_q = Sales_Vals[0:4]
cash_in_q = collections[0:4]
cash_out_q = np.abs(tot_disb[0:4])
ending_cash_q = end_cash[0:4]

# Membangun Dictionary Master (Kolom = Nama Akun, Baris = Periode Waktu)
master_data_dict = {
    'Period': periods,

    # --- QUANTITY / VOLUME DATA ---
    'Sales Volume (Units)': pad_yearly_sum(sales_units_q),
    # 'Production Volume (Units)': pad_yearly_sum(production_units_q),

    # --- OPERATIONAL BUDGETS ---
    'Revenue (Sales)': pad_yearly_sum(revenue_q),
    'Direct Material Purchased': pad_yearly_sum(Purchases_Vals[0:4]),
    'Direct Labor Cost': pad_yearly_sum(disb_dl[0:4]),
    'Overhead Cost': pad_yearly_sum(disb_oh[0:4]),
    'S&A Expense': pad_yearly_sum(disb_sa[0:4]),

    # --- INCOME STATEMENT (Data Tahunan) ---
    'Cost of Goods Sold (COGS)': pad_annual_only(COGS_Year),
    'Gross Margin': pad_annual_only(gross_margin),
    'Operating Income': pad_annual_only(op_income),
    'Interest Expense': pad_annual_only(interest_val),
    'Income Taxes': pad_annual_only(tax_val),
    'Net Income': pad_annual_only(net_income),

    # --- CASH BUDGET ---
    'Beginning Cash Balance': [beg_cash[0], beg_cash[1], beg_cash[2], beg_cash[3], beg_cash[0]], # Awal FY = Awal Q1
    'Total Cash Inflow': pad_yearly_sum(cash_in_q),
    'Total Cash Outflow': pad_yearly_sum(cash_out_q),
    'Ending Cash Balance': pad_yearly_last(ending_cash_q),

    # --- SOFP / BALANCE SHEET (Data Tahunan) ---
    'Current Assets': pad_annual_only(total_current_assets),
    'Non-Current Assets': pad_annual_only(total_non_current_assets),
    'Total Assets': pad_annual_only(total_assets),
    'Current Liabilities': pad_annual_only(current_liabilities),
    'Total Liabilities': pad_annual_only(total_liabilities),
    'Total Equity': pad_annual_only(total_equity),

    # --- FINANCIAL RATIOS & ANALYSIS (Data Tahunan) ---
    'Current Ratio (x)': pad_annual_only(current_ratio),
    'Quick Ratio (x)': pad_annual_only(quick_ratio),
    'Gross Profit Margin (%)': pad_annual_only(gross_profit_margin),
    'Net Profit Margin (%)': pad_annual_only(net_profit_margin),
    'ROA (%)': pad_annual_only(roa),
    'ROE (%)': pad_annual_only(roe),
    'Asset Turnover (x)': pad_annual_only(asset_turnover),
    'Altman Z-Score': pad_annual_only(z_score)
}

# Membuat DataFrame
df_master_dashboard = pd.DataFrame(master_data_dict)

# Menambahkan kolom status sebagai 'Projected'
df_master_dashboard['Status'] = 'Projected Budget'

# Menggeser kolom 'Status' ke urutan kedua (Setelah 'Period')
cols = ['Period', 'Status'] + [c for c in df_master_dashboard.columns if c not in ['Period', 'Status']]
df_master_dashboard = df_master_dashboard[cols]

# ====================================================================
# 2. PREVIEW & EXPORT KE EXCEL DALAM 1 SHEET
# ====================================================================
print("=== 🌟 MASTER DATASET DASHBOARD: WIDE FORMAT (1 SHEET) 🌟 ===")
# Menampilkan preview dengan format angka dipisah koma dan NaN ditampilkan sebagai strip (-)
display(df_master_dashboard.style.format(na_rep="-", precision=2).set_properties(**{'text-align': 'center'}))

# Menyimpan ke dalam Excel HANYA dengan 1 Sheet saja sesuai permintaan
file_name = 'Golden_Taste_Master_Dataset.xlsx'
df_master_dashboard.to_excel(file_name, index=False, sheet_name='Master_Data')

print(f"\n✅ Berhasil! Seluruh data budget (general) telah diekspor menjadi 1 sheet di file: {file_name}")

=== 🌟 MASTER DATASET DASHBOARD: WIDE FORMAT (1 SHEET) 🌟 ===


,Period,Status,Sales Volume (Units),Revenue (Sales),Direct Material Purchased,Direct Labor Cost,Overhead Cost,S&A Expense,Cost of Goods Sold (COGS),Gross Margin,Operating Income,Interest Expense,Income Taxes,Net Income,Beginning Cash Balance,Total Cash Inflow,Total Cash Outflow,Ending Cash Balance,Current Assets,Non-Current Assets,Total Assets,Current Liabilities,Total Liabilities,Total Equity,Current Ratio (x),Quick Ratio (x),Gross Profit Margin (%),Net Profit Margin (%),ROA (%),ROE (%),Asset Turnover (x),Altman Z-Score
0,Q1,Projected Budget,1000.00,10000.00,4492.00,1272.00,2281.00,1820.00,-,-,-,-,-,-,5200.00,10600.00,16466.60,1023.40,-,-,-,-,-,-,-,-,-,-,-,-,-,-
1,Q2,Projected Budget,1200.00,12000.00,5176.00,1512.00,2401.00,1940.00,-,-,-,-,-,-,1023.40,11850.00,10892.20,1611.20,-,-,-,-,-,-,-,-,-,-,-,-,-,-
2,Q3,Projected Budget,1500.00,15000.00,6480.00,1920.00,2605.00,2570.00,-,-,-,-,-,-,1611.20,14775.00,13314.20,3762.00,-,-,-,-,-,-,-,-,-,-,-,-,-,-
3,Q4,Projected Budget,2000.00,20000.00,6904.00,2160.00,2725.00,2320.00,-,-,-,-,-,-,3762.00,19625.00,17493.20,6583.80,-,-,-,-,-,-,-,-,-,-,-,-,-,-
4,FY (Full Year),Projected Budget,5700.00,57000.00,23052.00,6864.00,10012.00,8650.00,39617.00,17383.00,8733.00,60.00,3469.00,5204.00,5200.00,56850.00,58166.20,6583.80,9898.00,29840.00,39738.00,1381.00,1380.80,38357.00,7.17,5.85,0.30,0.09,0.13,0.14,1.43,20.44



✅ Berhasil! Seluruh data budget (general) telah diekspor menjadi 1 sheet di file: Golden_Taste_Master_Dataset.xlsx


# **Download Dataset in CSV File Version**

In [ ]:
import pandas as pd
import numpy as np
from IPython.display import display

# ====================================================================
# 1. PERSIAPAN MASTER DATASET (SINGLE SHEET WIDE FORMAT)
# ====================================================================
# Menggabungkan seluruh akun dari Sales Budget hingga Financial Analysis
# ke dalam satu format tabel mendatar (Wide Format) dengan baris sebagai Periode

periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY (Full Year)']

# --- Helper Functions untuk Padding Data (Menyesuaikan panjang array menjadi 5) ---
# Menjumlahkan Q1-Q4 untuk mendapatkan total FY (Cocok untuk Sales, Biaya, Cashflow)
def pad_yearly_sum(quarterly_array):
    return list(quarterly_array) + [sum(quarterly_array)]

# Mengambil nilai Q4 untuk mendapatkan total FY (Cocok untuk Ending Balance Kas / Piutang)
def pad_yearly_last(quarterly_array):
    return list(quarterly_array) + [quarterly_array[-1]]

# Membuat nilai Q1-Q4 kosong/NaN karena data hanya dihitung tahunan (Cocok untuk SOFP & Ratios)
def pad_annual_only(annual_value):
    return [np.nan, np.nan, np.nan, np.nan, annual_value]


# --- Ekstraksi Data Operasional Kuartalan ---
# Pastikan array ini ditarik dari variabel yang sudah sukses berjalan di atas
revenue_q = Sales_Vals[0:4]
cash_in_q = collections[0:4]
cash_out_q = np.abs(tot_disb[0:4])
ending_cash_q = end_cash[0:4]

# Membangun Dictionary Master (Kolom = Nama Akun, Baris = Periode Waktu)
master_data_dict = {
    'Period': periods,

    # --- OPERATIONAL BUDGETS ---
    'Revenue (Sales)': pad_yearly_sum(revenue_q),
    'Direct Material Purchased': pad_yearly_sum(Purchases_Vals[0:4]),
    'Direct Labor Cost': pad_yearly_sum(disb_dl[0:4]),
    'Overhead Cost': pad_yearly_sum(disb_oh[0:4]),
    'S&A Expense': pad_yearly_sum(disb_sa[0:4]),

    # --- INCOME STATEMENT (Data Tahunan) ---
    'Cost of Goods Sold (COGS)': pad_annual_only(COGS_Year),
    'Gross Margin': pad_annual_only(gross_margin),
    'Operating Income': pad_annual_only(op_income),
    'Interest Expense': pad_annual_only(interest_val),
    'Income Taxes': pad_annual_only(tax_val),
    'Net Income': pad_annual_only(net_income),

    # --- CASH BUDGET ---
    'Beginning Cash Balance': [beg_cash[0], beg_cash[1], beg_cash[2], beg_cash[3], beg_cash[0]], # Awal FY = Awal Q1
    'Total Cash Inflow': pad_yearly_sum(cash_in_q),
    'Total Cash Outflow': pad_yearly_sum(cash_out_q),
    'Ending Cash Balance': pad_yearly_last(ending_cash_q),

    # --- SOFP / BALANCE SHEET (Data Tahunan) ---
    'Current Assets': pad_annual_only(total_current_assets),
    'Non-Current Assets': pad_annual_only(total_non_current_assets),
    'Total Assets': pad_annual_only(total_assets),
    'Current Liabilities': pad_annual_only(current_liabilities),
    'Total Liabilities': pad_annual_only(total_liabilities),
    'Total Equity': pad_annual_only(total_equity),

    # --- FINANCIAL RATIOS & ANALYSIS (Data Tahunan) ---
    'Current Ratio (x)': pad_annual_only(current_ratio),
    'Quick Ratio (x)': pad_annual_only(quick_ratio),
    'Gross Profit Margin (%)': pad_annual_only(gross_profit_margin),
    'Net Profit Margin (%)': pad_annual_only(net_profit_margin),
    'ROA (%)': pad_annual_only(roa),
    'ROE (%)': pad_annual_only(roe),
    'Asset Turnover (x)': pad_annual_only(asset_turnover),
    'Altman Z-Score': pad_annual_only(z_score)
}

# Membuat DataFrame
df_master_dashboard = pd.DataFrame(master_data_dict)

# Menambahkan kolom status sebagai 'Projected'
df_master_dashboard['Status'] = 'Projected Budget'

# Menggeser kolom 'Status' ke urutan kedua (Setelah 'Period')
cols = ['Period', 'Status'] + [c for c in df_master_dashboard.columns if c not in ['Period', 'Status']]
df_master_dashboard = df_master_dashboard[cols]

# ====================================================================
# 2. PREVIEW & EXPORT KE EXCEL DALAM 1 SHEET
# ====================================================================
print("=== 🌟 MASTER DATASET DASHBOARD: WIDE FORMAT (1 SHEET) 🌟 ===")
# Menampilkan preview dengan format angka dipisah koma dan NaN ditampilkan sebagai strip (-)
display(df_master_dashboard.style.format(na_rep="-", precision=2).set_properties(**{'text-align': 'center'}))

# Menyimpan ke dalam Excel HANYA dengan 1 Sheet saja sesuai permintaan
file_name = 'Golden_Taste_Master_Dataset.xlsx'
df_master_dashboard.to_excel(file_name, index=False, sheet_name='Master_Data')

# Menyimpan ke dalam CSV
csv_file_name = 'Golden_Taste_Master_Dataset.csv'
df_master_dashboard.to_csv(csv_file_name, index=False)

print(f"\n✅ Berhasil! Seluruh data budget (general) telah diekspor menjadi 1 sheet di file: {file_name}")
print(f"✅ Berhasil! Seluruh data budget (general) juga telah diekspor ke file CSV: {csv_file_name}")

=== 🌟 MASTER DATASET DASHBOARD: WIDE FORMAT (1 SHEET) 🌟 ===


,Period,Status,Revenue (Sales),Direct Material Purchased,Direct Labor Cost,Overhead Cost,S&A Expense,Cost of Goods Sold (COGS),Gross Margin,Operating Income,Interest Expense,Income Taxes,Net Income,Beginning Cash Balance,Total Cash Inflow,Total Cash Outflow,Ending Cash Balance,Current Assets,Non-Current Assets,Total Assets,Current Liabilities,Total Liabilities,Total Equity,Current Ratio (x),Quick Ratio (x),Gross Profit Margin (%),Net Profit Margin (%),ROA (%),ROE (%),Asset Turnover (x),Altman Z-Score
0,Q1,Projected Budget,10000.00,4492.00,1272.00,2281.00,1820.00,-,-,-,-,-,-,5200.00,10600.00,16466.60,1023.40,-,-,-,-,-,-,-,-,-,-,-,-,-,-
1,Q2,Projected Budget,12000.00,5176.00,1512.00,2401.00,1940.00,-,-,-,-,-,-,1023.40,11850.00,10892.20,1611.20,-,-,-,-,-,-,-,-,-,-,-,-,-,-
2,Q3,Projected Budget,15000.00,6480.00,1920.00,2605.00,2570.00,-,-,-,-,-,-,1611.20,14775.00,13314.20,3762.00,-,-,-,-,-,-,-,-,-,-,-,-,-,-
3,Q4,Projected Budget,20000.00,6904.00,2160.00,2725.00,2320.00,-,-,-,-,-,-,3762.00,19625.00,17493.20,6583.80,-,-,-,-,-,-,-,-,-,-,-,-,-,-
4,FY (Full Year),Projected Budget,57000.00,23052.00,6864.00,10012.00,8650.00,39617.00,17383.00,8733.00,60.00,3469.00,5204.00,5200.00,56850.00,58166.20,6583.80,9898.00,29840.00,39738.00,1381.00,1380.80,38357.00,7.17,5.85,0.30,0.09,0.13,0.14,1.43,20.44



✅ Berhasil! Seluruh data budget (general) telah diekspor menjadi 1 sheet di file: Golden_Taste_Master_Dataset.xlsx
✅ Berhasil! Seluruh data budget (general) juga telah diekspor ke file CSV: Golden_Taste_Master_Dataset.csv


# **Dashboard for looker studio**

In [ ]:
import pandas as pd
import numpy as np
from IPython.display import display

# ====================================================================
# 1. PERSIAPAN MASTER DATASET (SINGLE SHEET WIDE FORMAT)
# ====================================================================
# Menggabungkan seluruh akun dari Sales Budget hingga Financial Analysis
# ke dalam satu format tabel mendatar (Wide Format) dengan baris sebagai Periode

periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY (Full Year)']

# --- Helper Functions untuk Padding Data (Menyesuaikan panjang array menjadi 5) ---
# Menjumlahkan Q1-Q4 untuk mendapatkan total FY (Cocok untuk Sales, Biaya, Cashflow)
def pad_yearly_sum(quarterly_array):
    return list(quarterly_array) + [sum(quarterly_array)]

# Mengambil nilai Q4 untuk mendapatkan total FY (Cocok untuk Ending Balance Kas / Piutang)
def pad_yearly_last(quarterly_array):
    return list(quarterly_array) + [quarterly_array[-1]]

# Membuat nilai Q1-Q4 kosong/NaN karena data hanya dihitung tahunan (Cocok untuk SOFP & Ratios)
def pad_annual_only(annual_value):
    return [np.nan, np.nan, np.nan, np.nan, annual_value]


# --- Ekstraksi Data Kuantitas (Volume) ---
# Menarik nilai 'Unit' penjualan dari df_Sales_Budget (Kolom Q1-Q4)
sales_units_q = df_Sales_Budget.loc['Unit'].iloc[0:4].values.flatten().astype(float)
# Catatan: Jika kamu ingin memasukkan Unit Produksi, kamu bisa un-comment baris di bawah ini:
# production_units_q = df_Production_Budget.iloc[-1, 0:4].values.flatten().astype(float)


# --- Ekstraksi Data Operasional Kuartalan ---
# Pastikan array ini ditarik dari variabel yang sudah sukses berjalan di atas
revenue_q = Sales_Vals[0:4]
cash_in_q = collections[0:4]
cash_out_q = np.abs(tot_disb[0:4])
ending_cash_q = end_cash[0:4]

# Membangun Dictionary Master (Kolom = Nama Akun, Baris = Periode Waktu)
master_data_dict = {
    'Period': periods,

    # --- QUANTITY / VOLUME DATA ---
    'Sales Volume (Units)': pad_yearly_sum(sales_units_q),
    # 'Production Volume (Units)': pad_yearly_sum(production_units_q),

    # --- OPERATIONAL BUDGETS ---
    'Revenue (Sales)': pad_yearly_sum(revenue_q),
    'Direct Material Purchased': pad_yearly_sum(Purchases_Vals[0:4]),
    'Direct Labor Cost': pad_yearly_sum(disb_dl[0:4]),
    'Overhead Cost': pad_yearly_sum(disb_oh[0:4]),
    'S&A Expense': pad_yearly_sum(disb_sa[0:4]),

    # --- INCOME STATEMENT (Data Tahunan) ---
    'Cost of Goods Sold (COGS)': pad_annual_only(COGS_Year),
    'Gross Margin': pad_annual_only(gross_margin),
    'Operating Income': pad_annual_only(op_income),
    'Interest Expense': pad_annual_only(interest_val),
    'Income Taxes': pad_annual_only(tax_val),
    'Net Income': pad_annual_only(net_income),

    # --- CASH BUDGET ---
    'Beginning Cash Balance': [beg_cash[0], beg_cash[1], beg_cash[2], beg_cash[3], beg_cash[0]], # Awal FY = Awal Q1
    'Total Cash Inflow': pad_yearly_sum(cash_in_q),
    'Total Cash Outflow': pad_yearly_sum(cash_out_q),
    'Ending Cash Balance': pad_yearly_last(ending_cash_q),

    # --- SOFP / BALANCE SHEET (Data Tahunan) ---
    'Current Assets': pad_annual_only(total_current_assets),
    'Non-Current Assets': pad_annual_only(total_non_current_assets),
    'Total Assets': pad_annual_only(total_assets),
    'Current Liabilities': pad_annual_only(current_liabilities),
    'Total Liabilities': pad_annual_only(total_liabilities),
    'Total Equity': pad_annual_only(total_equity),

    # --- FINANCIAL RATIOS & ANALYSIS (Data Tahunan) ---
    'Current Ratio (x)': pad_annual_only(current_ratio),
    'Quick Ratio (x)': pad_annual_only(quick_ratio),
    'Gross Profit Margin (%)': pad_annual_only(gross_profit_margin),
    'Net Profit Margin (%)': pad_annual_only(net_profit_margin),
    'ROA (%)': pad_annual_only(roa),
    'ROE (%)': pad_annual_only(roe),
    'Asset Turnover (x)': pad_annual_only(asset_turnover),
    'Altman Z-Score': pad_annual_only(z_score)
}

# Membuat DataFrame
df_master_dashboard = pd.DataFrame(master_data_dict)

# Menambahkan kolom status sebagai 'Projected'
df_master_dashboard['Status'] = 'Projected Budget'

# Menggeser kolom 'Status' ke urutan kedua (Setelah 'Period')
cols = ['Period', 'Status'] + [c for c in df_master_dashboard.columns if c not in ['Period', 'Status']]
df_master_dashboard = df_master_dashboard[cols]

# ====================================================================
# 2. PREVIEW & EXPORT KE EXCEL DALAM 1 SHEET
# ====================================================================
print("=== 🌟 MASTER DATASET DASHBOARD: WIDE FORMAT (1 SHEET) 🌟 ===")
# Menampilkan preview dengan format angka dipisah koma dan NaN ditampilkan sebagai strip (-)
display(df_master_dashboard.style.format(na_rep="-", precision=2).set_properties(**{'text-align': 'center'}))

# Menyimpan ke dalam Excel HANYA dengan 1 Sheet saja sesuai permintaan
file_name = 'Golden_Taste_Master_Dataset.xlsx'
df_master_dashboard.to_excel(file_name, index=False, sheet_name='Master_Data')

print(f"\n✅ Berhasil! Seluruh data budget (general) telah diekspor menjadi 1 sheet di file: {file_name}")

=== 🌟 MASTER DATASET DASHBOARD: WIDE FORMAT (1 SHEET) 🌟 ===


,Period,Status,Sales Volume (Units),Revenue (Sales),Direct Material Purchased,Direct Labor Cost,Overhead Cost,S&A Expense,Cost of Goods Sold (COGS),Gross Margin,Operating Income,Interest Expense,Income Taxes,Net Income,Beginning Cash Balance,Total Cash Inflow,Total Cash Outflow,Ending Cash Balance,Current Assets,Non-Current Assets,Total Assets,Current Liabilities,Total Liabilities,Total Equity,Current Ratio (x),Quick Ratio (x),Gross Profit Margin (%),Net Profit Margin (%),ROA (%),ROE (%),Asset Turnover (x),Altman Z-Score
0,Q1,Projected Budget,1000.00,10000.00,4492.00,1272.00,2281.00,1820.00,-,-,-,-,-,-,5200.00,10600.00,16466.60,1023.40,-,-,-,-,-,-,-,-,-,-,-,-,-,-
1,Q2,Projected Budget,1200.00,12000.00,5176.00,1512.00,2401.00,1940.00,-,-,-,-,-,-,1023.40,11850.00,10892.20,1611.20,-,-,-,-,-,-,-,-,-,-,-,-,-,-
2,Q3,Projected Budget,1500.00,15000.00,6480.00,1920.00,2605.00,2570.00,-,-,-,-,-,-,1611.20,14775.00,13314.20,3762.00,-,-,-,-,-,-,-,-,-,-,-,-,-,-
3,Q4,Projected Budget,2000.00,20000.00,6904.00,2160.00,2725.00,2320.00,-,-,-,-,-,-,3762.00,19625.00,17493.20,6583.80,-,-,-,-,-,-,-,-,-,-,-,-,-,-
4,FY (Full Year),Projected Budget,5700.00,57000.00,23052.00,6864.00,10012.00,8650.00,39617.00,17383.00,8733.00,60.00,3469.00,5204.00,5200.00,56850.00,58166.20,6583.80,9898.00,29840.00,39738.00,1381.00,1380.80,38357.00,7.17,5.85,0.30,0.09,0.13,0.14,1.43,20.44



✅ Berhasil! Seluruh data budget (general) telah diekspor menjadi 1 sheet di file: Golden_Taste_Master_Dataset.xlsx


In [ ]:
import pandas as pd
import numpy as np

# Calculate quarterly COGS based on Sales Units and Unit_Cost
# df_Sales_Unit contains quarterly sales units. Unit_Cost is the calculated unit manufacturing cost.
cogs_q_calculated = df_Sales_Unit.iloc[0, 0:4].values * Unit_Cost

# Create df_quarterly with the calculated COGS
df_quarterly = pd.DataFrame({
    'COGS': cogs_q_calculated
}, index=Quarter_Cols) # Quarter_Cols is assumed to be defined as ['1', '2', '3', '4']

# Define the quarterly cost variables from previously calculated DataFrames/arrays
# df_quarterly contains quarterly COGS
cogs_q = df_quarterly['COGS'].values
# Total_SA_Cost is a 2D numpy array of quarterly S&A costs
sa_cost_q = Total_SA_Cost[0]
# Total_Direct_Labor_cost is a 2D numpy array of quarterly direct labor costs
dl_cost_q = Total_Direct_Labor_cost[0]
# Total_Overhead_Cost is a 2D numpy array of quarterly overhead costs
oh_cost_q = Total_Overhead_Cost[0]

# Mengambil total nilai Full Year (FY) dari variabel yang sudah dihitung sebelumnya
# Asumsi: variabel-variabel ini (cogs_q, sa_cost_q, dll) adalah array/list 4 kuartal
total_cogs = cogs_q.sum()
total_sa = sa_cost_q.sum()
total_labor = dl_cost_q.sum()
total_overhead = oh_cost_q.sum()
total_interest = 60.0 * 4 # Asumsi interest 60 per kuartal sesuai script sebelumnya

# Membuat DataFrame dengan format Long (Memanjang ke bawah)
data_cost = {
    'Cost Category': [
        'Cost of Goods Sold',
        'S&A Expense',
        'Direct Labor Cost',
        'Overhead Cost',
        'Interest Expense'
    ],
    'Amount': [
        total_cogs,
        total_sa,
        total_labor,
        total_overhead,
        total_interest
    ]
}

df_cost_composition = pd.DataFrame(data_cost)

# Export ke Excel/CSV
df_cost_composition.to_excel('Dataset_Cost_Composition.xlsx', index=False)

print("Tabel Cost Composition berhasil dibuat!")
display(df_cost_composition)

Tabel Cost Composition berhasil dibuat!


,Cost Category,Amount
0,Cost of Goods Sold,39615.0
1,S&A Expense,8650.0
2,Direct Labor Cost,6864.0
3,Overhead Cost,10012.0
4,Interest Expense,240.0


In [ ]:
import pandas as pd

# ==============================================================================
# ASUMSI VARIABEL:
# Sesuaikan nama variabel di bawah ini dengan hasil perhitungan SOFP-mu sebelumnya.
# Jika kamu menggunakan df_sofp, kamu bisa menggunakan fungsi .loc[] seperti di bawah.
# Jika menggunakan variabel langsung (seperti sofp_ca, sofp_nca), tinggal ganti saja.
# ==============================================================================

# 1. Mendefinisikan Nilai Asset
current_assets_val = df_SOFP.loc['Total Current Asset', 'Amount']       # Atau pakai variabelmu: misal sofp_ca
non_current_assets_val = df_SOFP.loc['Total Non Current Asset', 'Amount'] # Atau pakai variabelmu: misal sofp_nca

# 2. Mendefinisikan Nilai Liabilities & Equity
total_liabilities_val = df_SOFP.loc['Total Liabilities', 'Amount']      # Atau pakai variabelmu: misal sofp_ap
total_equity_val = df_SOFP.loc['Total Equity', 'Amount']                # Atau pakai variabelmu: misal sofp_te

# ==============================================================================
# MEMBUAT DATAFRAME GABUNGAN (UNTUK KOMPOSISI KESELURUHAN)
# ==============================================================================

# Data untuk Komposisi Aset, Liabilitas, dan Ekuitas digabungkan
data_combined_sofp = {
    'Category': [
        'Current Assets',
        'Non-Current Assets',
        'Total Liabilities',
        'Total Equity'
    ],
    'Amount': [
        current_assets_val,
        non_current_assets_val,
        total_liabilities_val,
        total_equity_val
    ]
}
df_combined_sofp = pd.DataFrame(data_combined_sofp)

# ==============================================================================
# EXPORT KE EXCEL (1 File, 1 Sheet)
# ==============================================================================
file_name = 'Dataset_SOFP_Composition.xlsx'

with pd.ExcelWriter(file_name) as writer:
    df_combined_sofp.to_excel(writer, sheet_name='SOFP_Composition', index=False)

print(f"File '{file_name}' saved successfully to one sheet!")

File 'Dataset_SOFP_Composition.xlsx' saved successfully to one sheet!


# **Download file perhitungan budgeting bentuk excel**

In [ ]:
import pandas as pd
import numpy as np
import os
from openpyxl.styles import Font, Alignment, numbers

# Asumsi bahwa DataFrames berikut sudah terdefinisi di lingkungan Anda:
# df_Sales_Budget, df_Production_Budget, df_Material_A_Budget, df_Direct_Material_B_Budget,
# df_Direct_Labor_Budget, df_Overhead_Budget, df_Budgeted_SA, df_Income_Statement_Budget,
# df_Cash_Receive, df_Cash_Payment, df_Cash_Budget, df_SOFP, df_Financial_Analysis, df_Z_Score

# Daftar tabel yang akan diekspor, disesuaikan dengan variabel global yang ada.
# Saya telah menambahkan Material B dan memisahkan Financial Analysis menjadi dua bagian.
semua_tabel = [
    ("1. Sales Budget", df_Sales_Budget),
    ("2. Production Budget", df_Production_Budget),
    ("3. Direct Material Budget (Material A)", df_Direct_Material_A_Budget),
    ("4. Direct Material Budget (Material B)", df_Direct_Material_B_Budget),
    ("5. Direct Labor Budget", df_Direct_Labor_Budget),
    ("6. Manufacturing Overhead Budget", df_Overhead_Budget),
    ("7. Selling & Admin Budget", df_Budgeted_SA), # Menggunakan df_Budgeted_SA (variabel global)
    ("8. Budgeted Income Statement", df_Income_Statement_Budget), # Menggunakan df_Income_Statement_Budget (variabel global)
    ("9. Schedule of Cash Receipts", df_Cash_Receive), # Menggunakan df_Cash_Receive (variabel global)
    ("10. Schedule of Cash Payments", df_Cash_Payment),
    ("11. Cash Budget", df_Cash_Budget),
    ("12. Budgeted SOFP (Neraca)", df_SOFP),
    ("13. Financial Analysis", df_Financial_Analysis), # Menambahkan Financial Analysis
    ("14. Altman Z-Score Analysis", df_Z_Score)
]

output_excel_filename = 'Master_Budget_Report_Single_Sheet.xlsx'

print("Mendesain file Excel (Satu Sheet Lengkap) sesuai permintaan...")

with pd.ExcelWriter(output_excel_filename, engine='openpyxl') as writer:
    workbook = writer.book
    worksheet = workbook.create_sheet('Master_Budget_Report')
    writer.sheets['Master_Budget_Report'] = worksheet
    worksheet.sheet_view.showGridLines = False

    current_row = 1
    for title, df in semua_tabel:
        # Tulis Judul Tabel
        worksheet.cell(row=current_row, column=1, value=title).font = Font(bold=True, size=12)
        current_row += 1 # Pindah ke baris berikutnya untuk DataFrame

        # Tentukan label indeks untuk DataFrame
        index_label = 'Account / Indicator' if title.startswith(('12.', '13.', '14.')) else 'Account'

        # Tulis DataFrame ke lembar kerja. Header DataFrame akan dimulai dari `current_row`.
        df.to_excel(writer, sheet_name='Master_Budget_Report', startrow=current_row, startcol=0, index_label=index_label)

        # Terapkan format tebal pada baris header DataFrame (yaitu baris `current_row`)
        for col_idx in range(1, len(df.columns) + 2): # Meliputi kolom indeks dan semua kolom data
            worksheet.cell(row=current_row, column=col_idx).font = Font(bold=True)

        # Pindahkan kursor baris untuk tabel berikutnya
        # current_row + len(df) + 1 (untuk baris header DF + baris data DF)
        # + 3 (untuk 3 baris kosong antar tabel)
        current_row += len(df) + 1 + 3

    # Rapikan lebar kolom
    worksheet.column_dimensions['A'].width = 45
    for col_letter in ['B', 'C', 'D', 'E', 'F', 'G']:
        worksheet.column_dimensions[col_letter].width = 15

    # Formatting Angka (Koma ribuan dan Desimal)
    # Iterasi dimulai dari baris pertama (Excel) dan kolom B hingga G
    for row in worksheet.iter_rows(min_row=1, max_row=worksheet.max_row, min_col=2, max_col=7):
        for cell in row:
            if isinstance(cell.value, (int, float)):
                # Jika angka memiliki nilai absolut kecil (misal: rasio desimal) DAN memiliki desimal,
                # format dengan 3 angka desimal. Selain itu, gunakan pemisah ribuan tanpa desimal.
                if abs(cell.value) < 100 and (cell.value % 1 != 0):
                    cell.number_format = '0.000'
                else:
                    cell.number_format = numbers.FORMAT_NUMBER_COMMA_SEPARATED1

    # Hapus default 'Sheet' bawaan openpyxl jika ada
    if 'Sheet' in workbook.sheetnames:
        workbook.remove(workbook['Sheet'])

print(f"✅ BERHASIL! File telah direplikasi ke dalam satu Sheet.")
print(f"📂 Nama File: {os.path.abspath(output_excel_filename)}")

Mendesain file Excel (Satu Sheet Lengkap) sesuai permintaan...
✅ BERHASIL! File telah direplikasi ke dalam satu Sheet.
📂 Nama File: /content/Master_Budget_Report_Single_Sheet.xlsx
